#**Installing Email Validator**

In [ ]:
!pip install deep-translator email-validator

In [ ]:
!pip install googletrans==4.0.0-rc1


# **Import Libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from googletrans import Translator
from deep_translator import GoogleTranslator
from email_validator import validate_email, EmailNotValidError

#**Read the whole excel file containing the 4 sheets**


*   Sales_Orders_Raw
*   Products_Raw
*   Customers_Raw
*   Governorates_Lookup_Noise
*   Data_Dictionary





In [ ]:
dataSheets = pd.ExcelFile('/content/EG_Retail_Sales_Raw_CaseStudy 1(1).xlsx')
print(dataSheets.sheet_names)

['Sales_Orders_Raw', 'Products_Raw', 'Governorates_Lookup_Noise', 'Customers_Raw', 'Data_Dictionary']


# **Accessing Sales Orders Raw Sheet**

---



In [ ]:
SalesDf = dataSheets.parse('Sales_Orders_Raw')

# **Primary Inspection DF head,tail**




In [ ]:
SalesDf.head()

OrderID    OrderDate DeliveryDate CustomerID     CustomerName Gender  \
0  SO93810   2024-01-29   02/02/2024    CUS-575  Mohamed Mahmoud      M   
1  SO59823   12/03/2024   2024-03-15      C1077     سارة إبراهيم    ذكر   
2  SO11504     06-23-24  02 Jul 2024      C1020     كريم إبراهيم      M   
3  SO77839          NaN   01/05/2024      C1028       أحمد محمود      M   
4  SO80855  04 Jan 2024  08 Jan 2024    CUS-521      Alaa Rashid    NaN   

            Phone                    Email Governorate        City  ...  \
0             NaN    mohamed.mahmoud@gmail       Asyut    Manfalut  ...   
1  +2001284752529  سارة.إبراهyمexample.com       cairo  Heliopolis  ...   
2             NaN                      NaN       أسيوط      Abnoub  ...   
3             NaN     aحمد.محمود@gmail.com       أسوان    Kom Ombo  ...   
4   011 4828 5503                      NaN     RED SEA      Safaga  ...   

  PaymentMethod PaymentStatus ShipperName ShippingCost   SalesRep    Channel  \
0         Meeza          Paid      Aramex          NaN  سارة محمد      E-com   
1          Visa        Unpaid    بريد مصر        100.0   Karim R.      Store   
2          فوري         مدفوع         DHL         25.0   Karim R.      E-com   
3    MasterCard         مدفوع         DHL         25.0  Ahmed Ali  Tel-Sales   
4    MasterCard       Pending         DHL          0.0  سارة محمد   WhatsApp   

   ReturnFlag  ReturnDate     Status                            Notes  
0           N         NaN        NaN            Wrong phone / invalid  
1           N         NaN  Delivered  عنوان غير واضح - landmark: مسجد  
2           N         NaN       ملغي                     Deliver ASAP  
3           N         NaN        New                              NaN  
4           N         NaN       ملغي  عنوان غير واضح - landmark: مسجد  

[5 rows x 32 columns]

In [ ]:
SalesDf.tail()

OrderID    OrderDate DeliveryDate CustomerID   CustomerName Gender  \
145  SO94587   20/01/2024     02-03-24      C1029  Ahmed Mahmoud    ذكر   
146  SO26822     03-19-24  30 Mar 2024    CUS-596      هاني راشد      M   
147  SO24515   03/02/2024   05/02/2024        NaN       منى راشد   أنثى   
148  SO38476          NaN     05-12-24      C1052       Hany Ali      M   
149  SO16259  11 Jun 2024  22 Jun 2024      C1006   Alaa Mahmoud    NaN   

               Phone                     Email  Governorate              City  \
145  +20 01150629797   ahmed.mahmoud@gmail.com  South Sinai   Sharm El Sheikh   
146   +2001214323017       هانy.راشد@gmail.com     Qalyubia  Shubra El Kheima   
147      01137072082        منy.راشد@gmail.com     Dakahlia          Mansoura   
148    012 7199 5701       hany.aliexample.com  south sinai             Dahab   
149   +2001142016340  alaa.mahmoud@example.com      القاهرة            Shobra   

     ... PaymentMethod PaymentStatus ShipperName ShippingCost SalesRep  \
145  ...          Visa         مدفوع    بريد مصر         75.0    A.Ali   
146  ...         Meeza     غير مدفوع      Aramex         75.0   Mona-S   
147  ...          Cash         مدفوع      Aramex         75.0    A.Ali   
148  ...          Cash       Pending         DHL          0.0   Mona-S   
149  ...         Meeza       Pending      Aramex         25.0   Mona-S   

       Channel  ReturnFlag  ReturnDate   Status                         Notes  
145  Tel-Sales           N         NaN  Shipped          اتصل بنا قبل التوصيل  
146      Store           N         NaN  Shipped         عميل قديم - خصم إضافي  
147      E-com           N         NaN  Shipped  Customer asked for gift wrap  
148   WhatsApp           N         NaN     ملغي         Wrong phone / invalid  
149      E-com           N         NaN  Shipped                  Deliver ASAP  

[5 rows x 32 columns]


**Sales Column**


In [ ]:
SalesDf.columns

Index(['OrderID', 'OrderDate', 'DeliveryDate', 'CustomerID', 'CustomerName',
       'Gender', 'Phone', 'Email', 'Governorate', 'City', 'Address',
       'Latitude', 'Longitude', 'ProductSKU', 'ProductName', 'Category',
       'UnitPrice', 'Quantity', 'Subtotal', 'Discount', 'TotalAmount',
       'Currency', 'PaymentMethod', 'PaymentStatus', 'ShipperName',
       'ShippingCost', 'SalesRep', 'Channel', 'ReturnFlag', 'ReturnDate',
       'Status', 'Notes'],
      dtype='object')

**Checking Nulls and Data Types**

In [ ]:
SalesDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 32 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OrderID        150 non-null    object 
 1   OrderDate      135 non-null    object 
 2   DeliveryDate   137 non-null    object 
 3   CustomerID     96 non-null     object 
 4   CustomerName   150 non-null    object 
 5   Gender         134 non-null    object 
 6   Phone          123 non-null    object 
 7   Email          118 non-null    object 
 8   Governorate    150 non-null    object 
 9   City           150 non-null    object 
 10  Address        150 non-null    object 
 11  Latitude       130 non-null    object 
 12  Longitude      130 non-null    object 
 13  ProductSKU     115 non-null    object 
 14  ProductName    150 non-null    object 
 15  Category       150 non-null    object 
 16  UnitPrice      150 non-null    float64
 17  Quantity       150 non-null    int64  
 18  Subtotal  

In [ ]:
SalesDf.isna().sum()[SalesDf.isna().sum() > 1]

OrderDate        15
DeliveryDate     13
CustomerID       54
Gender           16
Phone            27
Email            32
Latitude         20
Longitude        20
ProductSKU       35
Subtotal         55
ShippingCost     17
ReturnDate      146
Status           16
Notes            30
dtype: int64

#**Conclusion**


*   Mixed Date formats in both of OrderDate and DeliveryDate

*   List item

*   List item

*   List item
*   List item


*   List item


*   List item


*   List item


 Mixed date formats (YYYY-MM-DD, DD/MM/YYYY, MM-DD-YY, Arabic month text)
 Inconsistent geography (Arabic/English/typos), swapped or out-of-range lat/long,
decimal commas
 Product/Category typos (e.g., “Electrnics”, “إلكترونيات”), SKU format variants/missing
 Quantity of 0 or negative; mismatched Subtotal vs. UnitPrice×Quantity
 Discount as value / percent string / decimal rate
 Currency as code/symbol or full value string (including USD)
 Invalid or partial Email, irregular Phone formats (+20, spaced, missing)
 Free-text notes with Arabic/English; multi-line addresses (embedded newlines)

# **Adressing each problem one by one -> my goal is to find a generic solution not hard coded**


# **Starting with Mixed Dates**

In [ ]:
def parse_mixed_date(date_str):
    date_str = str(date_str).strip()
    formats = [
        '%Y-%m-%d', '%d/%m/%Y', '%m/%d/%Y',
        '%d-%m-%Y', '%m-%d-%y', '%d %b %Y', '%d مارس %Y'
    ]
    for fmt in formats:
        try:
            return pd.to_datetime(date_str, format=fmt)
        except ValueError:
            pass


for idx, value in SalesDf['OrderDate'].items():
    parsed_date = parse_mixed_date(value)
    if parsed_date is not None and not pd.isna(parsed_date):
        SalesDf.loc[idx, 'OrderDate'] = parsed_date
    else:
        SalesDf.loc[idx, 'OrderDate'] = pd.NaT


In [ ]:
SalesDf['OrderDate'].isna().sum()

np.int64(15)

In [ ]:
SalesDf['OrderDate'].unique()

array([Timestamp('2024-01-29 00:00:00'), Timestamp('2024-03-12 00:00:00'),
       Timestamp('2024-06-23 00:00:00'), NaT,
       Timestamp('2024-01-04 00:00:00'), Timestamp('2024-04-01 00:00:00'),
       Timestamp('2024-03-29 00:00:00'), Timestamp('2024-02-20 00:00:00'),
       Timestamp('2024-03-25 00:00:00'), Timestamp('2024-01-11 00:00:00'),
       Timestamp('2024-01-23 00:00:00'), Timestamp('2024-05-30 00:00:00'),
       Timestamp('2024-01-02 00:00:00'), Timestamp('2024-03-26 00:00:00'),
       Timestamp('2024-05-10 00:00:00'), Timestamp('2024-05-04 00:00:00'),
       Timestamp('2024-01-28 00:00:00'), Timestamp('2024-04-03 00:00:00'),
       Timestamp('2024-06-13 00:00:00'), Timestamp('2024-01-19 00:00:00'),
       Timestamp('2024-02-27 00:00:00'), Timestamp('2024-06-02 00:00:00'),
       Timestamp('2024-02-07 00:00:00'), Timestamp('2024-04-16 00:00:00'),
       Timestamp('2024-01-18 00:00:00'), Timestamp('2024-05-07 00:00:00'),
       Timestamp('2024-06-12 00:00:00'), Timestamp('20

In [ ]:
for idx, value in SalesDf['DeliveryDate'].items():
    parsed_date = parse_mixed_date(value)
    if parsed_date is not None and not pd.isna(parsed_date):
        SalesDf.loc[idx, 'DeliveryDate'] = parsed_date
    else:
        SalesDf.loc[idx, 'DeliveryDate'] = pd.NaT


In [ ]:
SalesDf['OrderDate'].isna().sum()

np.int64(15)

# **Customer ID**

In [ ]:
SalesDf["CustomerID"].unique()

array(['CUS-575', 'C1077', 'C1020', 'C1028', 'CUS-521', 'C1016', 'C1079',
       'C1024', nan, 'CUS-527', 'CUS-563', 'C1097', 'CUS-500', 'CUS-550',
       'CUS-541', 'CUS-549', 'CUS-576', 'C1055', 'C1019', 'CUS-569',
       'C1053', 'CUS-564', 'CUS-580', 'C1043', 'C1049', 'C1099', 'C1089',
       'C1063', 'CUS-512', 'CUS-516', 'C1044', 'CUS-506', 'CUS-535',
       'CUS-533', 'C1042', 'C1001', 'C1070', 'C1065', 'C1034', 'CUS-519',
       'C1045', 'C1026', 'CUS-511', 'CUS-595', 'CUS-501', 'CUS-504',
       'CUS-520', 'C1041', 'C1000', 'C1031', 'CUS-539', 'CUS-538',
       'CUS-513', 'CUS-542', 'CUS-596', 'CUS-568', 'C1058', 'C1011',
       'CUS-582', 'C1032', 'C1062', 'C1022', 'CUS-560', 'C1035', 'C1093',
       'C1002', 'CUS-566', 'C1094', 'CUS-515', 'CUS-565', 'CUS-523',
       'C1018', 'CUS-598', 'CUS-562', 'C1098', 'C1080', 'C1071', 'C1005',
       'C1029', 'C1052', 'C1006'], dtype=object)

In [ ]:
SalesDf['CustomerID'] = SalesDf['CustomerID'].apply(lambda x: x.replace('CUS-', 'C') if isinstance(x, str) and x.startswith('CUS-') else x)


In [ ]:
SalesDf["CustomerID"].unique()

array(['C575', 'C1077', 'C1020', 'C1028', 'C521', 'C1016', 'C1079',
       'C1024', nan, 'C527', 'C563', 'C1097', 'C500', 'C550', 'C541',
       'C549', 'C576', 'C1055', 'C1019', 'C569', 'C1053', 'C564', 'C580',
       'C1043', 'C1049', 'C1099', 'C1089', 'C1063', 'C512', 'C516',
       'C1044', 'C506', 'C535', 'C533', 'C1042', 'C1001', 'C1070',
       'C1065', 'C1034', 'C519', 'C1045', 'C1026', 'C511', 'C595', 'C501',
       'C504', 'C520', 'C1041', 'C1000', 'C1031', 'C539', 'C538', 'C513',
       'C542', 'C596', 'C568', 'C1058', 'C1011', 'C582', 'C1032', 'C1062',
       'C1022', 'C560', 'C1035', 'C1093', 'C1002', 'C566', 'C1094',
       'C515', 'C565', 'C523', 'C1018', 'C598', 'C562', 'C1098', 'C1080',
       'C1071', 'C1005', 'C1029', 'C1052', 'C1006'], dtype=object)

#**Customer Name**


*   We have a Mix of Arabic & English names


      1.  Either we hard code it
      2.   List item





In [ ]:
SalesDf["CustomerName"].unique()

array(['Mohamed Mahmoud', 'سارة إبراهيم', 'كريم إبراهيم', 'أحمد محمود',
       'Alaa Rashid', 'يوسف راشد', 'Alaa Mahmoud', 'Ahmed Farouk',
       'يوسف إبراهيم', 'Sara Farouk', 'كريم علي', 'Dina Hassan',
       'Youssef Rashid', 'نور فتحي', 'محمد فتحي', 'Ahmed Fathy',
       'محمد محمود', 'Sara Mahmoud', 'هاني راشد', 'هاني علي', 'Mona Nasr',
       'Nour Fathy', 'نور علي', 'Mona Ibrahim', 'منى إبراهيم', 'يوسف حسن',
       'دينا محمود', 'يوسف نصر', 'Hany Rashid', 'Dina Ibrahim',
       'علاء فاروق', 'علاء فتحي', 'كريم حسن', 'Alaa Farouk',
       'Nour Farouk', 'Nour Mahmoud', 'منى علي', 'Ahmed Ibrahim',
       'علاء حسن', 'سارة فتحي', 'Alaa Nasr', 'منى محمود', 'Hany Ibrahim',
       'نور فاروق', 'Mohamed Nasr', 'علاء إبراهيم', 'Nour Hassan',
       'محمد علي', 'Hany Ali', 'Alaa Hassan', 'Mona Hassan', 'Hany Fathy',
       'يوسف علي', 'أحمد نصر', 'Mohamed Rashid', 'سارة محمود',
       'Ahmed Mahmoud', 'Ahmed Ali', 'Mona Fathy', 'نور راشد',
       'Dina Fathy', 'سارة حسن', 'علاء علي', 'No

In [ ]:
translator = Translator()
translated = translator.translate('سارة إبراهيم', src='ar', dest='en')
print(translated.text)

Sarah Ibrahim


#**After seeing the result on the sample case I'll apply it to the customer name column**

In [ ]:
SalesDf['CustomerName'] = SalesDf['CustomerName'].apply(
    lambda x: translator.translate(x, src='ar', dest='en').text if isinstance(x, str) else x
)

In [ ]:
SalesDf["CustomerName"].unique()

array(['Mohamed Mahmoud', 'Sarah Ibrahim', 'Karim Ibrahim',
       'Ahmed Mahmoud', 'Alaa Rashid', 'Youssef Rashid', 'Alaa Mahmoud',
       'Ahmed Farouk', 'Youssef Ibrahim', 'Sarah Farouk', 'Karim Ali',
       'Dina Hassan', 'Nour Fathi', 'Muhammad Fathi', 'Ahmed Fathi',
       'Muhammad Mahmoud', 'Sarah Mahmoud', 'Hani Rashid', 'Hani Ali',
       'Mona Nasr', 'Nour Ali', 'Mona Ibrahim', 'Youssef Hassan',
       'Dina Mahmoud', 'Youssef Nasr', 'Hany Rashid', 'Dina Ibrahim',
       'Alaa Farouk', 'Alaa Fathi', 'Karim Hassan', 'Nour Farouk',
       'Nour Mahmoud', 'Mona Ali', 'Ahmed Ibrahim', 'Alaa Hassan',
       'Sarah Fathi', 'Alaa Nasr', 'Mona Mahmoud', 'Hany Ibrahim',
       'Mohamed Nasr', 'Alaa Ibrahim', 'Nour Hassan', 'Muhammad Ali',
       'Hany Ali', 'Mona Hassan', 'Hany Fathi', 'Yusuf Ali', 'Ahmed Nasr',
       'Mohamed Rashid', 'Ahmed Ali', 'Mona Fathi', 'Nour Rashid',
       'Dina Fathi', 'Sarah Hassan', 'Alaa Ali', 'Muhammad Hassan',
       'Hani Nasr', 'Youssef Ali', 'Ahm

#**Now we Don't have any Arabic Entries in the Customer Name**

In [ ]:
SalesDf.columns

Index(['OrderID', 'OrderDate', 'DeliveryDate', 'CustomerID', 'CustomerName',
       'Gender', 'Phone', 'Email', 'Governorate', 'City', 'Address',
       'Latitude', 'Longitude', 'ProductSKU', 'ProductName', 'Category',
       'UnitPrice', 'Quantity', 'Subtotal', 'Discount', 'TotalAmount',
       'Currency', 'PaymentMethod', 'PaymentStatus', 'ShipperName',
       'ShippingCost', 'SalesRep', 'Channel', 'ReturnFlag', 'ReturnDate',
       'Status', 'Notes'],
      dtype='object')

# **Phone Normalization**
**irregular Phone formats (+20, spaced, missing)**

In [ ]:
SalesDf["Phone"].unique()

array([nan, '+2001284752529', '011 4828 5503', '01065337219',
       '+2001067062156', '01051746937', '+20 01190014570',
       '+2001249209916', '+2001289160218', '+20 01275007200',
       '+20 01016172803', '+20 01091478884', '01239711402',
       '+20 01210380581', '+20 01034215035', '+20 01284990607',
       '+2001252431444', '01054250344', '+2001158895439', '012 2862 8439',
       '01098195059', '011 3317 3544', '01248124318', '01285383550',
       '01042507072', '01048983895', '+20 01011243818', '+2001192493631',
       '01073360324', '+2001050502136', '+2001112058055', '01216511491',
       '01019865900', '+20 01063108878', '011 8514 1654',
       '+2001128830282', '+2001092698414', '+20 01076399516',
       '01121989814', '+2001152476240', '+20 01076627415',
       '011 1925 1149', '+2001198911718', '010 3659 2965',
       '+2001146297263', '+20 01080638102', '010 6556 8711',
       '01023017361', '+2001038439415', '011 4616 0937', '011 9745 4571',
       '012 8175 7508', '+200

In [ ]:
SalesDf['Phone'] = SalesDf['Phone'].apply(lambda x: str(x).replace('+20', '').strip() if pd.notnull(x) else x)

In [ ]:
SalesDf["Phone"].unique()

array([nan, '01284752529', '011 4828 5503', '01065337219', '01067062156',
       '01051746937', '01190014570', '01249209916', '01289160218',
       '01275007200', '01016172803', '01091478884', '01239711402',
       '01210380581', '01034215035', '01284990607', '01252431444',
       '01054250344', '01158895439', '012 2862 8439', '01098195059',
       '011 3317 3544', '01248124318', '01285383550', '01042507072',
       '01048983895', '01011243818', '01192493631', '01073360324',
       '01050502136', '01112058055', '01216511491', '01019865900',
       '01063108878', '011 8514 1654', '01128830282', '01092698414',
       '01076399516', '01121989814', '01152476240', '01076627415',
       '011 1925 1149', '01198911718', '010 3659 2965', '01146297263',
       '01080638102', '010 6556 8711', '01023017361', '01038439415',
       '011 4616 0937', '011 9745 4571', '012 8175 7508', '01087346846',
       '01293635734', '01220072800', '010 6520 1639', '01271231190',
       '012 6162 8545', '0108263488

#**All phone numbers are now without prefix**

#**Gender**

In [ ]:
SalesDf["Gender"].unique()

array(['M', 'ذكر', nan, 'أنثى', 'F'], dtype=object)

In [ ]:
(SalesDf['Gender'].astype(str).str.strip().str.lower() == 'nan').sum()

np.int64(16)

In [ ]:
SalesDf['Gender'] = SalesDf['Gender'].astype(str).str.strip().str.lower().map({'m':'male', 'f':'female', 'ذكر':'male', 'أنثى':'female'})

In [ ]:
SalesDf["Gender"].unique()


array(['male', nan, 'female'], dtype=object)

In [ ]:
genderNulls=SalesDf[SalesDf.isna().any(axis=1)]['CustomerName']
print(genderNulls)

0      Mohamed Mahmoud
1        Sarah Ibrahim
2        Karim Ibrahim
3        Ahmed Mahmoud
4          Alaa Rashid
            ...       
145      Ahmed Mahmoud
146        Hani Rashid
147        Mona Rashid
148           Hany Ali
149       Alaa Mahmoud
Name: CustomerName, Length: 150, dtype: object


#**I'll fix the gender by myself as I can get it from the name**

In [ ]:
NameGender= {
    'alaa rashid': 'male',
    'sarah ibrahim': 'female',
    'karim ali': 'male',
    'sarah farouk': 'female',
    'alaa farouk': 'male',
    'dina ibrahim': 'female',
    'mona fathi': 'female',
    'nour farouk': 'female',
    'ahmed ibrahim': 'male',
    'mohamed ali': 'male',
    'mona ali': 'female',
    'dina ali': 'female',
    'ahmed rashid': 'male',
    'muhammad mahmoud': 'male',
    'alaa mahmoud': 'male'
}

#**Mapping**

In [ ]:
SalesDf['Gender'] = SalesDf.apply(
    lambda row: NameGender.get(str(row['CustomerName']).strip().lower(), row['Gender']),
    axis=1
)

#**All the Nulls in the Gender Column and the inconsistancies are fixed**

In [ ]:
SalesDf["Gender"].isna().sum()

np.int64(0)

#**Inconsistent geography (Arabic/English/typos)**

In [ ]:
SalesDf["Governorate"].unique()

array(['Asyut', 'cairo', 'أسيوط', 'أسوان', 'RED SEA', 'Red Sea',
       'south sinai', 'جنوب سيناء', 'asyut', 'giza', 'الجيزة', 'SHARQIA',
       'GHARBIA', 'البحر الأحمر', 'South Sinai', 'Alexandria', 'sharqia',
       'القليوبية', 'Sharqia', 'GIZA', 'Al Red Sea', 'luxor', 'Luxor',
       'SOUTH SINAI', 'ASYUT', 'LUXOR', 'Al Alexandria', 'الإسكندرية',
       'Al Gharbia', 'Giza', 'Dakahlia', 'الأقصر', 'الغربية', 'aswan',
       'QALYUBIA', 'qalyubia', 'alexandria', 'القاهرة', 'Gharbia',
       'القاهره', 'Gizah', 'Cairo', 'الدقهلية', 'dakahlia', 'Qalyubia',
       'DAKAHLIA', 'Aswan', 'Al South Sinai', 'Al Luxor', 'الشرقية',
       'Al Giza', 'Al Asyut', 'Al Aswan'], dtype=object)

In [ ]:
SalesDf['Governorate'] = SalesDf['Governorate'].apply(
    lambda x: translator.translate(x, src='ar', dest='en').text if isinstance(x, str) else x)

In [ ]:
SalesDf["Governorate"].unique()

array(['Asyut', 'cairo', 'Assiut', 'Aswan', 'RED SEA', 'Red Sea',
       'south sinai', 'South Sinai', 'asyut', 'giza', 'Giza', 'SHARQIA',
       'GHARBIA', 'Alexandria', 'sharqia', 'Qalyubia', 'Sharqia', 'GIZA',
       'Al Red Sea', 'luxor', 'Luxor', 'SOUTH SINAI', 'ASYUT', 'LUXOR',
       'Al Alexandria', 'Al Gharbia', 'Dakahlia', 'Western', 'aswan',
       'QALYUBIA', 'qalyubia', 'alexandria', 'Cairo', 'Gharbia', 'Gizah',
       'Al South Sinai', 'Al Luxor', 'Eastern', 'Al Giza', 'Al Assiut',
       'Al Aswan'], dtype=object)

In [ ]:
SalesDf = SalesDf.applymap(
    lambda x: str(x).lower().replace('eastern', 'gharbia').replace('western','sharqia').replace('gizah','giza').replace('asyut','assiut').removeprefix('al ') if not str(x).lower().startswith('alexandria') else str(x).lower()
     if isinstance(x, str) else x
)


/tmp/ipython-input-1660195904.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  SalesDf = SalesDf.applymap(


#**The Inconsistancies in the Governorate Column are Fixed**

In [ ]:
SalesDf["Governorate"].unique()

array(['assiut', 'cairo', 'aswan', 'red sea', 'south sinai', 'giza',
       'sharqia', 'gharbia', 'alexandria', 'qalyubia', 'luxor',
       'dakahlia'], dtype=object)

#**Converted lat/long values from string to float, fixed decimal commas, flagged coordinates, and swapped the wrong latitude and longitude**

In [ ]:
SalesDf[['Latitude', 'Longitude']] = SalesDf[['Latitude', 'Longitude']].applymap(lambda x: float(str(x).replace(',', '.')) if pd.notna(x) else pd.NA)

/tmp/ipython-input-3483825830.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  SalesDf[['Latitude', 'Longitude']] = SalesDf[['Latitude', 'Longitude']].applymap(lambda x: float(str(x).replace(',', '.')) if pd.notna(x) else pd.NA)


In [ ]:
SalesDf[['Latitude', 'Longitude']]

Latitude  Longitude
0    31.183700  27.180000
1    30.085855  31.231585
2    27.217872  31.216867
3          NaN        NaN
4    27.293172  33.777343
..         ...        ...
145  28.543865  34.001116
146  31.280000  30.420000
147  31.019691  31.416599
148  28.499424  33.979843
149        NaN        NaN

[150 rows x 2 columns]

In [ ]:
SalesDf['LatLong_Swapped'] = ~SalesDf.apply(lambda row: -90 <= row['Latitude'] <= 90 and -180 <= row['Longitude'] <= 180, axis=1)

In [ ]:
SalesDf['LatLong_Swapped'].value_counts()

LatLong_Swapped
False    127
True      23
Name: count, dtype: int64

In [ ]:
SalesDf.loc[SalesDf['LatLong_Swapped'], ['Latitude', 'Longitude']] = SalesDf.loc[SalesDf['LatLong_Swapped'], ['Longitude', 'Latitude']].values

#**Invalid or partial Email**

In [ ]:
SalesDf["Email"].unique()

array(['mohamed.mahmoud@gmail', 'سارة.إبراهyمexample.com', 'nan',
       'aحمد.محمود@gmail.com', 'yوسف.راشد@example.com',
       'سارة.إبراهyم@gmail.com', 'aحمد.محمودexample.com',
       'alaa.mahmoud@gmail.com', 'كرyم.علy@gmail', 'dina.hassan@gmail',
       'youssef.rashid@gmail', 'نور.فتحyexample.com',
       'محمد.فتحy@example.com', 'ahmed.fathy@gmail.com',
       'محمد.محمود@gmail', 'sara.mahmoud@gmail.com',
       'هانy.راشد@gmail.com', 'هانy.علyexample.com',
       'mona.nasr@example.com', 'nour.fathy@gmail', 'mona.ibrahim@gmail',
       'منy.إبراهyم@gmail', 'yوسف.حسن@gmail', 'دyنا.محمود@example.com',
       'sara.faroukexample.com', 'yوسف.نصرexample.com',
       'علاء.فاروقexample.com', 'كرyم.حسن@gmail.com',
       'alaa.faroukexample.com', 'nour.mahmoud@gmail.com',
       'منy.علy@gmail', 'ahmed.ibrahim@example.com',
       'علاء.حسنexample.com', 'alaa.nasr@example.com', 'منy.محمود@gmail',
       'hany.ibrahimexample.com', 'نور.فاروق@gmail.com',
       'mohamed.nasrexample.com'

#**Function For Adjusting Email Formate**

In [ ]:
from email_validator import validate_email, EmailNotValidError
from deep_translator import GoogleTranslator

def smart_email_cleaner(email):
    if not isinstance(email, str):
        return None

    email = email.strip()

    # Fix common typos
    if 'example.com' in email and '@' not in email:
        email = email.replace('example.com', '@example.com')
    if 'gmail' in email and '@' not in email:
        email = email.replace('gmail', '@gmail.com')

    # Translate only the name part
    parts = email.split('@')
    if len(parts) == 2:
        local, domain = parts
        try:
            local_translated = GoogleTranslator(source='auto', target='en').translate(local)
            email = f"{local_translated}@{domain}"
        except:
            pass  # If translation fails, keep original

    # Validate
    try:
        return validate_email(email, check_deliverability=False).email
    except EmailNotValidError:
        return email  # Keep original if invalid

**Applying to DF**

In [ ]:
SalesDf['Email_Cleaned'] = SalesDf['Email'].apply(smart_email_cleaner)

In [ ]:
SalesDf["Email_Cleaned"].unique()

array(['mohamed. mahmoud@gmail', 'Sarah. Ibrahim@example.com', 'nan',
       'Ahmed. Mahmoud@gmail.com', 'Youssef Rashid@example.com',
       'Sarah. Ibrahim@gmail.com', 'Ahmed. Mahmoud@example.com',
       'alaa. mahmoud@gmail.com', 'Thank you@gmail', 'dina. hassan@gmail',
       'youssef.rashid@gmail', 'Light. Fathy@example.com',
       'Muhammad.Fathy@example.com', 'ahmed.fathy@gmail.com',
       'Muhammad. Mahmoud@gmail', 'sara.mahmoud@gmail.com',
       'Hany.Rashed@gmail.com', 'Hany. Aliy@example.com',
       'mona.nasr@example.com', 'nour.fathy@gmail', 'mona.ibrahim@gmail',
       'From Y. Ibrahim@gmail', 'Yousef. Hasan@gmail',
       'Dr. Na Mahmoud@example.com', 'sara.farouk@example.com',
       'Youssef Nasr@example.com', 'Alaa Farouk@example.com',
       'Generous. Hassan@gmail.com', 'alaa. farouk@example.com',
       'nour.mahmoud@gmail.com', 'My.Aly@gmail',
       'ahmed.ibrahim@example.com', 'Alaa. Hassan@example.com',
       'alaa.nasr@example.com', 'My. Mahmoud@gmail',


In [ ]:
#SalesDf['Email_Cleaned'] = SalesDf['Email_Cleaned'].apply(
#    lambda x: str(x).replace(' ', '').strip().lower() if isinstance(x, str) else x
#)

In [ ]:
SalesDf['Email_Cleaned'] = SalesDf['Email_Cleaned'].apply(lambda x: (str(x).replace(' ', '').lower() + '.com') if isinstance(x, str) and '@' in x and '.com' not in x else str(x).replace(' ', '').lower() if isinstance(x, str) else x)

In [ ]:
SalesDf["Email"]=SalesDf["Email_Cleaned"]

#**ALL the Emails Are Consistant and Valide**

In [ ]:
SalesDf["Email_Cleaned"].unique()

array(['mohamed.mahmoud@gmail.com', 'sarah.ibrahim@example.com', 'nan',
       'ahmed.mahmoud@gmail.com', 'youssefrashid@example.com',
       'sarah.ibrahim@gmail.com', 'ahmed.mahmoud@example.com',
       'alaa.mahmoud@gmail.com', 'thankyou@gmail.com',
       'dina.hassan@gmail.com', 'youssef.rashid@gmail.com',
       'light.fathy@example.com', 'muhammad.fathy@example.com',
       'ahmed.fathy@gmail.com', 'muhammad.mahmoud@gmail.com',
       'sara.mahmoud@gmail.com', 'hany.rashed@gmail.com',
       'hany.aliy@example.com', 'mona.nasr@example.com',
       'nour.fathy@gmail.com', 'mona.ibrahim@gmail.com',
       'fromy.ibrahim@gmail.com', 'yousef.hasan@gmail.com',
       'dr.namahmoud@example.com', 'sara.farouk@example.com',
       'youssefnasr@example.com', 'alaafarouk@example.com',
       'generous.hassan@gmail.com', 'alaa.farouk@example.com',
       'nour.mahmoud@gmail.com', 'my.aly@gmail.com',
       'ahmed.ibrahim@example.com', 'alaa.hassan@example.com',
       'alaa.nasr@example.co

#**Product/Category typos (e.g., “Electrnics”, “إلكترونيات”), SKU format variants/missing**

In [ ]:
SalesDf["ProductName"].unique()

array(['data warehousing 101', 'men t-shirt - تيشيرت',
       'air fryer 4l - قلاية', 'bluetooth سماعة',
       'puzzle 1000pcs - أحجية', 'organic زيت زيتون 1l',
       'laptop i7 16gb - لابتوب', 'smartphone a55 - هاتف',
       'football - كرة قدم', 'electric kettle - غلاية',
       'smartphone a55 - تليفون', 'labtop i7 16gb - لابتوب'], dtype=object)

In [ ]:
SalesDf["Category"].unique()

array(['books', 'fashion', 'home', 'electronics', 'toys', 'grocery',
       'electrnics', 'sports', 'إلكترونيات'], dtype=object)

In [ ]:
SalesDf['Category'] = SalesDf['Category'].apply(
    lambda x: 'electronics' if str(x).lower().strip() in ['electrnics', 'إلكترونيات', 'electronics'] else str(x).lower().strip()
)

In [ ]:
SalesDf["Category"].unique()

array(['books', 'fashion', 'home', 'electronics', 'toys', 'grocery',
       'sports'], dtype=object)

# **Standardize and Impute ProductSKU Values Based on ProductName**

In [ ]:
SalesDf['ProductSKU'].unique()

array(['book-900', 'fash-210', 'book900', 'home-102', 'nan', 'elec003',
       'toy-777', 'elec-001', 'elec-002', 'elec-003', 'fash210',
       'sprt301', 'home-101', 'toy777', 'home101', 'sprt-301', 'groc-501',
       'elec002', 'groc501', 'elec001', 'home102'], dtype=object)

**35 NULL entry in form of string**

In [ ]:
SalesDf['ProductSKU'].apply(lambda x: str(x).lower().strip() == 'nan').sum()


np.int64(35)

In [ ]:
SalesDf['ProductSKU'].apply(lambda x: x == 'book-900').sum()


np.int64(9)

#**Standardization**

---




In [ ]:
SalesDf['ProductSKU'] = SalesDf['ProductSKU'].apply(lambda x: str(x).replace('-', '').strip().lower() if pd.notna(x) else x)

In [ ]:
SalesDf['ProductSKU'].unique()

array(['book900', 'fash210', 'home102', 'nan', 'elec003', 'toy777',
       'elec001', 'elec002', 'sprt301', 'home101', 'groc501'],
      dtype=object)

**NULLS**
appears to have 0 nulls because they are strings 'nan' but the 'nan' appears in the .unique as entry in the column

In [ ]:
SalesDf['ProductSKU'] = SalesDf.groupby('ProductName')['ProductSKU'].transform(
    lambda group: group.replace('nan', np.nan).fillna(method='ffill').fillna(method='bfill')
)

/tmp/ipython-input-1409685333.py:2: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  lambda group: group.replace('nan', np.nan).fillna(method='ffill').fillna(method='bfill')


In [ ]:
SalesDf['ProductSKU'].unique()

array(['book900', 'fash210', 'home102', 'elec003', 'toy777', 'groc501',
       'elec001', 'elec002', 'sprt301', 'home101'], dtype=object)

# **ProductSKU have no issues left**

# **Currency**

In [ ]:
SalesDf['Currency'].unique()

array(['ج.م', 'egp', 'e£ 35564.57', 'e£', 'usd', 'usd 10930.9',
       'usd 1034.18', 'ج.م 67637.94', 'usd 1293.05', 'e£ 41982.116',
       'usd 51507.984000000004', 'egp 145935.24', 'e£ 1998.65',
       'e£ 95373.45', 'e£ 61003.33', 'usd 71287.912', 'usd 87.46',
       'e£ 38326.83', 'e£ 75509.0685', 'egp 98410.32', 'egp 7171.444',
       'ج.م 106755.28', 'ج.م 123945.77', 'egp 24877.57', 'egp 22540.86',
       'ج.م 10966.19', 'egp 56905.87', 'ج.م 115228.22999999998',
       'egp 163839.511', 'e£ 78208.7', 'ج.م 91116.16', 'egp 113751.165',
       'e£ 58102.682', 'ج.م 226644.3', 'ج.م 100.0', 'usd 255.61',
       'ج.م 103629.55', 'usd 51.79', 'e£ 17146.47', 'ج.م 40907.34',
       'ج.م 19213.17', 'egp 32678.7', 'ج.م 62899.26', 'egp 19878.56',
       'egp 8489.664', 'usd 941.31', 'usd 193036.37999999998',
       'egp 135644.08', 'usd 180172.4', 'egp 72897.65',
       'ج.م 114421.26400000001', 'e£ 27938.9', 'e£ 60411.76',
       'usd 10196.858', 'usd 1300.99', 'e£ 115758.61'], dtype=object)

#**Inspecting Unit Price**

In [ ]:
SalesDf.columns

Index(['OrderID', 'OrderDate', 'DeliveryDate', 'CustomerID', 'CustomerName',
       'Gender', 'Phone', 'Email', 'Governorate', 'City', 'Address',
       'Latitude', 'Longitude', 'ProductSKU', 'ProductName', 'Category',
       'UnitPrice', 'Quantity', 'Subtotal', 'Discount', 'TotalAmount',
       'Currency', 'PaymentMethod', 'PaymentStatus', 'ShipperName',
       'ShippingCost', 'SalesRep', 'Channel', 'ReturnFlag', 'ReturnDate',
       'Status', 'Notes', 'LatLong_Swapped', 'Email_Cleaned'],
      dtype='object')

In [ ]:
SalesDf['UnitPrice'].unique()

array(['30206.12', '37363.34', '43906.58', '23377.88', '29772.0',
       '4660.31', '11185.03', '20333.45', '22530.98', '4243.18',
       '33092.99', '15344.99', '4304.79', '29921.93', '35503.65',
       '24791.67', '11647.81', '39336.31', '44276.42', '26327.97',
       '4706.84', '44172.94', '40076.04', '30769.67', '32192.49',
       '40072.9', '31965.64', '2238.12', '1788.21', '18076.78',
       '41270.55', '9532.21', '7489.59', '49906.24', '33508.21',
       '39548.84', '3951.23', '36406.51', '34830.6', '23326.9',
       '26494.41', '37625.11', '27871.4', '48389.77', '19587.15',
       '18321.36', '1683.12', '49624.69', '49592.32', '43005.13',
       '31447.11', '24832.57', '46623.26', '22505.86', '2579.35',
       '38214.28', '18950.29', '17123.32', '2278.62', '5670.32',
       '8728.18', '33875.95', '21052.6', '31353.01', '33141.02',
       '42607.34', '25303.1', '35119.7', '34346.23', '10194.21',
       '5780.36', '26764.98', '32251.49', '45319.86', '25857.31',
       '43371.75',

#**Transform Unit Price into Float instead of String to allow calculations**

In [ ]:
SalesDf['UnitPrice'] = SalesDf['UnitPrice'].astype('Float64')

#**Solving Quantity of 0 or negative; mismatched Subtotal vs. UnitPrice×Quantity**

#

 25  ShippingCost   133 non-null    float64
 26  SalesRep       150 non-null    object
 27  Channel        150 non-null    object
 28  ReturnFlag     150 non-null    object
 29  ReturnDate     4 non-null      object
 30  Status         134 non-null    object
 31  Notes  

# **Payment Method**

In [ ]:
SalesDf['PaymentMethod'].unique()

array(['meeza', 'visa', 'فوري', 'mastercard', 'cod', 'cash'], dtype=object)

In [ ]:
SalesDf['PaymentMethod'] = SalesDf['PaymentMethod'].replace('فوري', 'Fawry')

#**Standardize Arabic Payment Method to English:**

In [ ]:
SalesDf['PaymentMethod'].value_counts()

PaymentMethod
visa          29
cash          28
meeza         25
Fawry         25
cod           22
mastercard    21
Name: count, dtype: int64

#**Standardize Arabic Payment Status to English:**

In [ ]:
SalesDf['PaymentStatus'].unique()

array(['paid', 'unpaid', 'مدفوع', 'pending', 'غير مدفوع'], dtype=object)

In [ ]:
SalesDf['PaymentStatus'] = SalesDf['PaymentStatus'].str.strip().str.lower().replace({
    'مدفوع': 'paid',
    'غير مدفوع': 'unpaid'
})

In [ ]:
SalesDf['PaymentStatus'].value_counts()

PaymentStatus
unpaid     64
paid       56
pending    30
Name: count, dtype: int64

#**convert the Arabic in Shipper Name Column**

In [ ]:
SalesDf["ShipperName"].unique()

array(['aramex', 'بريد مصر', 'dhl', 'fedex'], dtype=object)

In [ ]:
SalesDf['ShipperName'] = SalesDf['ShipperName'].replace('بريد مصر', 'Egypt Post')

In [ ]:
SalesDf['ShipperName'].value_counts()

ShipperName
aramex        47
Egypt Post    40
dhl           32
fedex         31
Name: count, dtype: int64

In [ ]:
SalesDf["SalesRep"].unique()

array(['سارة محمد', 'karim r.', 'ahmed ali', 'أحمد علي', 'a.ali',
       'mona-s'], dtype=object)

#**Since they are only 4 Sales Representitives we can use the mapping method no need for translation**

In [ ]:
SalesDf['SalesRep'] = SalesDf['SalesRep'].str.strip().str.lower().replace({'سارة محمد':'sarah mohamed','karim r.':'karim rashid','أحمد علي':'ahmed ali','a.ali':'ahmed ali','mona-s':'mona salah'})

In [ ]:
SalesDf['SalesRep'].value_counts()

SalesRep
ahmed ali        74
karim rashid     32
mona salah       28
sarah mohamed    16
Name: count, dtype: int64

# **Replaced the Arabic term تجارة إلكترونية with its English translation e-com to have unified naming**

In [ ]:
SalesDf["Channel"].unique()

array(['e-com', 'store', 'tel-sales', 'whatsapp', 'تجارة إلكترونية'],
      dtype=object)

In [ ]:
SalesDf['Channel'] = SalesDf['Channel'].str.strip().str.lower().replace({'تجارة إلكترونية': 'e-com'})

In [ ]:
SalesDf['Channel'].value_counts()

Channel
e-com        65
tel-sales    31
store        29
whatsapp     25
Name: count, dtype: int64

#**Standardize the ReturnFlag column**

In [ ]:
SalesDf['ReturnFlag'].unique()

array(['n', 'لا', 'نعم', '0', 'no', '1', 'y'], dtype=object)

In [ ]:
SalesDf['ReturnFlag'].value_counts()

ReturnFlag
n      128
no       6
لا       5
نعم      5
0        4
1        1
y        1
Name: count, dtype: int64

In [ ]:
SalesDf['ReturnFlag'] = SalesDf['ReturnFlag'].str.strip().str.lower().replace({'نعم':'yes','y':'yes','1':'yes','لا':'no','n':'no','0':'no','no':'no'})

#**We only have 2 status yes, no**

In [ ]:
SalesDf['ReturnFlag'].value_counts()

ReturnFlag
no     143
yes      7
Name: count, dtype: int64

# **Return Date Unification**

In [ ]:
SalesDf['ReturnDate'].unique()

array(['nan', '22/06/2024', '21 مارس 2024', '12/04/2024', '13 may 2024'],
      dtype=object)

In [ ]:
SalesDf['ReturnDate'] = SalesDf['ReturnDate'].apply(
    lambda x: pd.Timestamp('2024-06-22') if str(x).strip() == '22/06/2024' else
              pd.Timestamp('2024-04-12') if str(x).strip() == '12/04/2024' else
              pd.Timestamp('2024-11-13') if str(x).strip() == '11 13' else
              pd.Timestamp('2024-05-01') if 'may' in str(x).lower() or 'مايو' in str(x) else
              pd.NaT
)

In [ ]:
SalesDf.columns

Index(['OrderID', 'OrderDate', 'DeliveryDate', 'CustomerID', 'CustomerName',
       'Gender', 'Phone', 'Email', 'Governorate', 'City', 'Address',
       'Latitude', 'Longitude', 'ProductSKU', 'ProductName', 'Category',
       'UnitPrice', 'Quantity', 'Subtotal', 'Discount', 'TotalAmount',
       'Currency', 'PaymentMethod', 'PaymentStatus', 'ShipperName',
       'ShippingCost', 'SalesRep', 'Channel', 'ReturnFlag', 'ReturnDate',
       'Status', 'Notes', 'LatLong_Swapped', 'Email_Cleaned'],
      dtype='object')

In [ ]:
SalesDf['ReturnDate'].unique()

<DatetimeArray>
['NaT', '2024-06-22 00:00:00', '2024-04-12 00:00:00', '2024-05-01 00:00:00']
Length: 4, dtype: datetime64[ns]

# **Standardize Order Status Labels**

In [ ]:
SalesDf['Status'].unique()

array(['nan', 'delivered', 'ملغي', 'new', 'shipped', 'processing',
       'returned'], dtype=object)

In [ ]:
SalesDf['Status'].value_counts()

Status
shipped       29
ملغي          27
processing    26
delivered     20
returned      17
nan           16
new           15
Name: count, dtype: int64

In [ ]:
SalesDf['Status'] = SalesDf['Status'].str.strip().str.lower().replace({'ملغي': 'cancelled'})

In [ ]:
SalesDf['Status'].value_counts()

Status
shipped       29
cancelled     27
processing    26
delivered     20
returned      17
nan           16
new           15
Name: count, dtype: int64

#**- Translates Arabic Notes to English using translator object**

In [ ]:
SalesDf["Notes"].value_counts()

Notes
nan                                30
عنوان غير واضح - landmark: مسجد    25
customer asked for gift wrap       22
عميل قديم - خصم إضافي              22
deliver asap                       20
wrong phone / invalid              16
اتصل بنا قبل التوصيل               15
Name: count, dtype: int64

In [ ]:
SalesDf['Notes'] = SalesDf['Notes'].apply(lambda x: translator.translate(x, src='ar', dest='en').text if isinstance(x, str) and any('\u0600' <= c <= '\u06FF' for c in x) else x)

In [ ]:
tr = GoogleTranslator(source='auto', target='en')
SalesDf['Notes'] = SalesDf['Notes'].apply(lambda x: tr.translate(x) if isinstance(x, str) else x)

In [ ]:
SalesDf["Notes"].value_counts()

Notes
in                                    30
customer asked for gift wrap          22
Old customer - extra discount         22
Unclear address - landmark: mosque    21
deliver asap                          20
wrong phone / invalid                 16
Contact us before delivery            15
Unclear title - landmark: Mosque       4
Name: count, dtype: int64

#**multi-line addresses (embedded newlines) & Dealing with arabic words**

In [ ]:
SalesDf["Address"].unique()

array(['88\napt 18\nblock 7', '146\napt 46\nعمارة 7', '139 شارع tahrir',
       '186 شارع الهرم', '192\napt 21\nblock 2', '28\napt 48\nblock 9',
       '172 شارع nile', '130 st. الهرم', '119\napt 43\nعمارة 25',
       '125 street الهرم', '20\napt 38\nblock 4', '143 street الهرم',
       '88 street tahrir', '19 street nile', '160 st. الهرم',
       '147\napt 25\nblock 28', '63 شارع الهرم', '80\napt 29\nblock 19',
       '36 street tahrir', '4\napt 23\nblock 19', '107\napt 47\nblock 11',
       '145 شارع الهرم', '148\napt 28\nعمارة 23', '130\napt 5\nعمارة 15',
       '150\napt 43\nblock 30', '80\napt 31\nblock 12',
       '141\napt 47\nعمارة 4', '200\napt 44\nعمارة 23',
       '71\napt 41\nblock 3', '77\napt 11\nblock 26', '33 شارع النصر',
       '13\napt 19\nعمارة 28', '143 st. nile', '106\napt 45\nblock 14',
       '55\napt 13\nعمارة 30', '73 st. الهرم', '96 st. الهرم',
       '161\napt 44\nعمارة 18', '102\napt 16\nعمارة 25',
       '96\napt 4\nعمارة 22', '122 st. tahrir', '93\napt 7\n

In [ ]:
SalesDf['Address'] = SalesDf['Address'].apply(lambda x: translator.translate(str(x).replace('\n', ' '), src='ar', dest='en').text if any('\u0600' <= c <= '\u06FF' for c in str(x)) else str(x).replace('\n', ' '))

#**Final Address Result**

In [ ]:
SalesDf['Address'].unique()

array(['88 apt 18 block 7', '146 apt 46 building 7', '139 Tahrir Street',
       '186 Haram Street', '192 apt 21 block 2', '28 apt 48 block 9',
       '172 nile street', '130 st.Pyramid', '119 apt 43 building 25',
       '125 street, Haram', '20 apt 38 block 4', '143 street Haram',
       '88 street tahrir', '19 street nile', '160 st.Pyramid',
       '147 apt 25 block 28', '63 Haram Street', '80 apt 29 block 19',
       '36 street tahrir', '4 apt 23 block 19', '107 apt 47 block 11',
       '145 Haram Street', '148 apt 28 building 23',
       '130 apt 5 building 15', '150 apt 43 block 30',
       '80 apt 31 block 12', '141 apt 47 building 4',
       '200 apt 44 building 23', '71 apt 41 block 3',
       '77 apt 11 block 26', '33 Al-Nasr Street', '13 apt 19 building 28',
       '143 st. nile', '106 apt 45 block 14', '55 apt 13 building 30',
       '73 st.Pyramid', '96 st.Pyramid', '161 apt 44 building 18',
       '102 apt 16 building 25', '96 apt 4 building 22', '122 st. tahrir',
       '

#**Discount as value / percent string / decimal rate**


In [ ]:
SalesDf["Discount"].value_counts()

Discount
20      25
0       16
15%     16
0.1     15
10%     14
0.05    13
5       13
15      13
0.2      9
5%       9
10       7
Name: count, dtype: int64

In [ ]:
SalesDf['Discount'] = SalesDf['Discount'].apply(lambda x: float(str(x).replace('%', '')) / 100 if '%' in str(x) else float(str(x)))

In [ ]:
SalesDf["Discount"].value_counts()

Discount
0.10     29
20.00    25
0.05     22
0.15     16
0.00     16
15.00    13
5.00     13
0.20      9
10.00     7
Name: count, dtype: int64

#**Standardized mixed discount formats by converting percentage strings, whole numbers, and decimals into decimal**

#**Solving Quantity of 0 or negative; mismatched Subtotal vs. UnitPrice×Quantity**

In [ ]:
SalesDf['Quantity'].unique()

array(['4', '5', '1', '2', '3', '-1', '0', '10'], dtype=object)

In [ ]:
SalesDf['Quantity'].apply(lambda x: x == '0').sum()


np.int64(4)

**Transform Quantity from object to Int**

In [ ]:
SalesDf[(SalesDf['Quantity'] == 0) | (SalesDf['Quantity'] == -1)][['Quantity', 'UnitPrice', 'Subtotal', 'ReturnFlag','TotalAmount']]

Empty DataFrame
Columns: [Quantity, UnitPrice, Subtotal, ReturnFlag, TotalAmount]
Index: []

In [ ]:
SalesDf['Quantity'] = SalesDf['Quantity'].astype('Int64')

In [ ]:
if 'ReturnFlag' not in SalesDf: SalesDf['ReturnFlag'] = 'no'
SalesDf['ReturnFlag'] = SalesDf['ReturnFlag'].astype(str).str.strip().str.lower()
SalesDf['Quantity'] = pd.to_numeric(SalesDf['Quantity'], errors='coerce')

SalesDf.loc[SalesDf['ReturnFlag']=='yes', 'Quantity'] = 0
SalesDf.loc[(SalesDf['ReturnFlag']!='yes') & (SalesDf['Quantity']==-1), 'Quantity'] = 1
SalesDf.loc[(SalesDf['ReturnFlag']!='yes') & (SalesDf['Quantity'].isna() | (SalesDf['Quantity']==0)), 'Quantity'] = 1



In [ ]:
SalesDf['Quantity'].unique()

<IntegerArray>
[4, 5, 1, 2, 3, 10, 0]
Length: 7, dtype: Int64

In [ ]:
SubTotalNew= SalesDf.eval("UnitPrice * Quantity")

/tmp/ipython-input-3887973959.py:1: RuntimeWarning: Engine has switched to 'python' because numexpr does not support extension array dtypes. Please set your engine to python manually.
  SubTotalNew= SalesDf.eval("UnitPrice * Quantity")


In [ ]:
for c in ['Quantity','UnitPrice','Subtotal']:
    SalesDf[c] = pd.to_numeric(
        SalesDf[c].astype(str).str.replace(r'[^0-9\.-]', '', regex=True),
        errors='coerce'
    )


In [ ]:
FillNew= SalesDf['Subtotal'].isna() | (~np.isclose(SalesDf['Subtotal'].fillna(SubTotalNew), SubTotalNew, atol=1e-6))

In [ ]:
print(SalesDf.loc[FillNew, ['Quantity','UnitPrice','Subtotal']].to_string(index=True))

     Quantity  UnitPrice   Subtotal
0           4   30206.12  121699.93
3           1   23377.88        NaN
4           2   29772.00        NaN
5           3    4660.31   12706.11
7           3   20333.45   58400.68
8           3   22530.98        NaN
9           4    4243.18   15892.34
10          1   33092.99  -29867.43
11          4   15344.99   62036.39
12          1    4304.79        NaN
15          1   24791.67  -26624.77
17          3   39336.31  118737.10
18          4   44276.42  181202.21
19          4   26327.97  104222.43
20          1    4706.84        NaN
21          1   44172.94   45709.88
23          4   30769.67        NaN
24          2   32192.49        NaN
25          5   40072.90  182419.05
26          1   31965.64   29477.80
27          1    2238.12    2155.58
28          1    1788.21        NaN
29          4   18076.78        NaN
30          1   41270.55        NaN
31         10    9532.21   95298.45
32          2    7489.59        NaN
34          2   33508.21   6

In [ ]:
SalesDf.loc[FillNew, 'Subtotal'] = SubTotalNew[FillNew]

#**Now they are Matching**

In [ ]:
print(SalesDf.loc[FillNew, ['Quantity','UnitPrice','Subtotal']].to_string(index=True))

     Quantity  UnitPrice   Subtotal
0           4   30206.12  120824.48
3           1   23377.88   23377.88
4           2   29772.00   59544.00
5           3    4660.31   13980.93
7           3   20333.45   61000.35
8           3   22530.98   67592.94
9           4    4243.18   16972.72
10          1   33092.99   33092.99
11          4   15344.99   61379.96
12          1    4304.79    4304.79
15          1   24791.67   24791.67
17          3   39336.31  118008.93
18          4   44276.42  177105.68
19          4   26327.97  105311.88
20          1    4706.84    4706.84
21          1   44172.94   44172.94
23          4   30769.67  123078.68
24          2   32192.49   64384.98
25          5   40072.90  200364.50
26          1   31965.64   31965.64
27          1    2238.12    2238.12
28          1    1788.21    1788.21
29          4   18076.78   72307.12
30          1   41270.55   41270.55
31         10    9532.21   95322.10
32          2    7489.59   14979.18
34          2   33508.21   6

#**Currency as code/symbol or full value string (including USD)**

In [ ]:
SalesDf["Currency"].unique()

array(['ج.م', 'egp', 'e£ 35564.57', 'e£', 'usd', 'usd 10930.9',
       'usd 1034.18', 'ج.م 67637.94', 'usd 1293.05', 'e£ 41982.116',
       'usd 51507.984000000004', 'egp 145935.24', 'e£ 1998.65',
       'e£ 95373.45', 'e£ 61003.33', 'usd 71287.912', 'usd 87.46',
       'e£ 38326.83', 'e£ 75509.0685', 'egp 98410.32', 'egp 7171.444',
       'ج.م 106755.28', 'ج.م 123945.77', 'egp 24877.57', 'egp 22540.86',
       'ج.م 10966.19', 'egp 56905.87', 'ج.م 115228.22999999998',
       'egp 163839.511', 'e£ 78208.7', 'ج.م 91116.16', 'egp 113751.165',
       'e£ 58102.682', 'ج.م 226644.3', 'ج.م 100.0', 'usd 255.61',
       'ج.م 103629.55', 'usd 51.79', 'e£ 17146.47', 'ج.م 40907.34',
       'ج.م 19213.17', 'egp 32678.7', 'ج.م 62899.26', 'egp 19878.56',
       'egp 8489.664', 'usd 941.31', 'usd 193036.37999999998',
       'egp 135644.08', 'usd 180172.4', 'egp 72897.65',
       'ج.م 114421.26400000001', 'e£ 27938.9', 'e£ 60411.76',
       'usd 10196.858', 'usd 1300.99', 'e£ 115758.61'], dtype=object)

In [ ]:
SalesDf['Currency'] = SalesDf['Currency'].apply(lambda x: 'usd' if 'usd' in str(x).lower() else 'egp' if any(c in str(x).lower() for c in ['ج.م', 'e£', 'egp']) else pd.NA)

#**The Currency only includes data about if it is Egp or USD because we already have the TotalAmount column no need for redundant numbers in the currency column**

In [ ]:
SalesDf['Currency'].value_counts()

Currency
egp    104
usd     46
Name: count, dtype: int64

#**DataFrame After Cleaning**

In [ ]:
SalesDf.head()

OrderID            OrderDate         DeliveryDate CustomerID  \
0  so93810  2024-01-29 00:00:00  2024-02-02 00:00:00       c575   
1  so59823  2024-03-12 00:00:00  2024-03-15 00:00:00      c1077   
2  so11504  2024-06-23 00:00:00  2024-07-02 00:00:00      c1020   
3  so77839                  nat  2024-05-01 00:00:00      c1028   
4  so80855  2024-01-04 00:00:00  2024-01-08 00:00:00       c521   

      CustomerName  Gender          Phone                      Email  \
0  mohamed mahmoud    male            nan  mohamed.mahmoud@gmail.com   
1    sarah ibrahim  female    01284752529  sarah.ibrahim@example.com   
2    karim ibrahim    male            nan                        nan   
3    ahmed mahmoud    male            nan    ahmed.mahmoud@gmail.com   
4      alaa rashid    male  011 4828 5503                        nan   

  Governorate        City  ... ShipperName  ShippingCost       SalesRep  \
0      assiut    manfalut  ...      aramex           nan  sarah mohamed   
1       cairo  heliopolis  ...  Egypt Post         100.0   karim rashid   
2      assiut      abnoub  ...         dhl          25.0   karim rashid   
3       aswan    kom ombo  ...         dhl          25.0      ahmed ali   
4     red sea      safaga  ...         dhl           0.0  sarah mohamed   

     Channel ReturnFlag ReturnDate     Status  \
0      e-com         no        NaT        nan   
1      store         no        NaT  delivered   
2      e-com         no        NaT  cancelled   
3  tel-sales         no        NaT        new   
4   whatsapp         no        NaT  cancelled   

                                Notes  LatLong_Swapped  \
0               wrong phone / invalid            False   
1  Unclear address - landmark: mosque            False   
2                        deliver asap            False   
3                                  in             True   
4  Unclear address - landmark: mosque            False   

               Email_Cleaned  
0  mohamed.mahmoud@gmail.com  
1  sarah.ibrahim@example.com  
2                        nan  
3    ahmed.mahmoud@gmail.com  
4                        nan  

[5 rows x 34 columns]

In [ ]:
print(SalesDf['OrderID'].duplicated().sum())

14


In [ ]:
print(SalesDf[SalesDf['OrderID'].duplicated(keep=False)])

     OrderID            OrderDate         DeliveryDate CustomerID  \
3    so77839                  nat  2024-05-01 00:00:00      c1028   
5    so19508                  nat  2024-01-03 00:00:00      c1016   
6    so15229  2024-04-01 00:00:00  2024-04-03 00:00:00      c1079   
7    so77839  2024-03-29 00:00:00  2024-04-08 00:00:00      c1024   
9    so15229  2024-03-25 00:00:00  2024-01-29 00:00:00        nan   
14   so63523  2024-03-26 00:00:00  2024-03-29 00:00:00      c1097   
19   so15229  2024-06-13 00:00:00  2024-06-13 00:00:00        nan   
22   so63523  2024-06-02 00:00:00  2024-06-06 00:00:00        nan   
25   so77839                  nat  2024-05-18 00:00:00       c576   
26   so13495  2024-04-16 00:00:00  2024-04-20 00:00:00        nan   
29   so19508  2024-05-07 00:00:00  2024-05-07 00:00:00        nan   
42   so52778  2024-01-08 00:00:00  2024-01-16 00:00:00        nan   
60   so13495  2024-06-25 00:00:00  2024-06-30 00:00:00       c535   
73   so57774  2024-03-21 00:00:00 

In [ ]:
SalesDf = SalesDf.sort_values(by=['OrderDate', 'CustomerID', 'Email', 'Phone'], ascending=[False, False, False, False]).drop_duplicates(subset='OrderID', keep='first')

In [ ]:
print(SalesDf['OrderID'].duplicated().sum())

0


#**Products_Raw**
---

In [ ]:
ProductDf = dataSheets.parse('Products_Raw')

# **Primary Inspection DF head,tail**



In [ ]:
ProductDf.head()

SKU              ProductName     Category
0  ELEC-001  Laptop i7 16GB - لابتوب  electronics
1  elec-001  LAPTOP I7 16GB - لابتوب   إلكترونيات
2  ELEC-001  Labtop i7 16GB - لابتوب  electronics
3  elec-002    Smartphone A55 - هاتف   Electrnics
4   ELEC002    Smartphone A55 - هاتف   إلكترونيات

In [ ]:
ProductDf.tail()

SKU             ProductName Category
25  HOME102    Air Fryer 4L - قلاية     home
26  HOME102    AIR FRYER 4L - قلاية     Home
27  TOY-777  PUZZLE 1000PCS - أحجية     Toys
28  TOY-777  PUZZLE 1000PCS - أحجية     Toys
29  TOY-777  Puzzle 1000pcs - أحجية     toys

In [ ]:
ProductDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   SKU          30 non-null     object
 1   ProductName  30 non-null     object
 2   Category     30 non-null     object
dtypes: object(3)
memory usage: 852.0+ bytes


In [ ]:
ProductDf.isna().sum()[ProductDf.isna().sum() > 1]

Series([], dtype: int64)

#**Product/Category typos (e.g., “Electrnics”, “إلكترونيات”), SKU format variants/missing**

In [ ]:
ProductDf["ProductName"].unique()

array(['Laptop i7 16GB - لابتوب', 'LAPTOP I7 16GB - لابتوب',
       'Labtop i7 16GB - لابتوب', 'Smartphone A55 - هاتف',
       'Electric Kettle - غلاية', 'ELECTRIC KETTLE - غلاية',
       'Organic زيت زيتون 1L', 'Men T-Shirt - تيشيرت',
       'FOOTBALL - كرة قدم', 'Football - كرة قدم', 'Data Warehousing 101',
       'DATA WAREHOUSING 101', 'BLUETOOTH سماعة', 'Bluetooth سماعة',
       'Air Fryer 4L - قلاية', 'AIR FRYER 4L - قلاية',
       'PUZZLE 1000PCS - أحجية', 'Puzzle 1000pcs - أحجية'], dtype=object)

In [ ]:
ProductDf["Category"].unique()

array(['electronics', 'إلكترونيات', 'Electrnics', 'Home', 'grocery',
       'Grocery', 'Fashion', 'sports', 'Books', 'books', 'home', 'Toys',
       'toys'], dtype=object)

In [ ]:
ProductDf['Category'] = ProductDf['Category'].apply(
    lambda x: 'electronics' if str(x).lower().strip() in ['electrnics', 'إلكترونيات', 'electronics'] else str(x).lower().strip()
)

In [ ]:
ProductDf["Category"].unique()

array(['electronics', 'home', 'grocery', 'fashion', 'sports', 'books',
       'toys'], dtype=object)

# **Standardize and Impute ProductSKU Values Based on ProductName**

In [ ]:
ProductDf['SKU'].unique()

array(['ELEC-001', 'elec-001', 'elec-002', 'ELEC002', 'home-101',
       'HOME101', 'groc-501', 'GROC501', 'GROC-501', 'fash-210',
       'FASH-210', 'sprt-301', 'SPRT-301', 'SPRT301', 'BOOK900',
       'book-900', 'ELEC003', 'ELEC-003', 'elec-003', 'HOME102',
       'TOY-777'], dtype=object)

**35 NULL entry in form of string**

In [ ]:
ProductDf['SKU'].apply(lambda x: str(x).lower().strip() == 'nan').sum()


np.int64(0)

In [ ]:
ProductDf['SKU'].apply(lambda x: x == 'book-900').sum()


np.int64(1)

#**Standardization**

---




In [ ]:
ProductDf['SKU'] = ProductDf['SKU'].apply(lambda x: str(x).replace('-', '').strip().lower() if pd.notna(x) else x)

In [ ]:
ProductDf['SKU'].unique()

array(['elec001', 'elec002', 'home101', 'groc501', 'fash210', 'sprt301',
       'book900', 'elec003', 'home102', 'toy777'], dtype=object)

**NULLS**
appears to have 0 nulls because they are strings 'nan' but the 'nan' appears in the .unique as entry in the column

In [ ]:
ProductDf['SKU'] = ProductDf.groupby('ProductName')['SKU'].transform(
    lambda group: group.replace('nan', np.nan).fillna(method='ffill').fillna(method='bfill')
)

/tmp/ipython-input-1427144408.py:2: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  lambda group: group.replace('nan', np.nan).fillna(method='ffill').fillna(method='bfill')


In [ ]:
ProductDf['SKU'].unique()

array(['elec001', 'elec002', 'home101', 'groc501', 'fash210', 'sprt301',
       'book900', 'elec003', 'home102', 'toy777'], dtype=object)

# **SKU have no issues left**

#**Customers_Raw**


# **Accessing Customers Raw Sheet**

---



In [ ]:
CustomersDf = dataSheets.parse('Customers_Raw')

# **Primary Inspection DF head,tail**




In [ ]:
CustomersDf.head()

CustomerID  CustomerName Gender            Phone                    Email  \
0      C1000    Alaa Fathy      M      01232297072    alaa.fathyexample.com   
1      C1001   Mohamed Ali      M      01288174504  mohamed.ali@example.com   
2    CUS-502  Youssef Nasr    ذكر  +20 01142158168       youssef.nasr@gmail   
3    CUS-503  علاء إبراهيم      M   +2001247509496  علاء.إبراهyمexample.com   
4        NaN      يوسف علي      M    012 5369 1252       yوسف.علy@gmail.com   

        Gov             City           Address  
0   Sharqia            Tanta   8 Street Tahrir  
1   gharbia  Sharm El Sheikh   134 شارع Tahrir  
2     LUXOR            Tanta  26 Street Tahrir  
3     luxor  Sharm El Sheikh  106 Street النصر  
4  Dakahlia        Nasr City    90 شارع Tahrir

In [ ]:
CustomersDf.tail()

CustomerID  CustomerName Gender            Phone                     Email  \
55        NaN  Karim Farouk    ذكر      01058384619  karim.farouk@example.com   
56        NaN  Karim Rashid      M  +20 01177008305   karim.rashidexample.com   
57    CUS-557   Dina Hassan   أنثى   +2001247376435   dina.hassan@example.com   
58        NaN   Alaa Farouk      F              NaN         alaa.farouk@gmail   
59    CUS-559      سارة حسن      M  +20 01012302458                       NaN   

              Gov      City            Address  
55  Al Alexandria     Tanta  154 Street Tahrir  
56         الجيزة  Hurghada      38 شارع النصر  
57     alexandria     Tanta     82 شارع Tahrir  
58        Gharbia     Maadi   99 Street Tahrir  
59     Alexandria     Tanta   93 Street Tahrir


**Customer Columns**


In [ ]:
CustomersDf.columns

Index(['CustomerID', 'CustomerName', 'Gender', 'Phone', 'Email', 'Gov', 'City',
       'Address'],
      dtype='object')

**Checking Nulls and Data Types**

In [ ]:
CustomersDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   CustomerID    40 non-null     object
 1   CustomerName  60 non-null     object
 2   Gender        54 non-null     object
 3   Phone         50 non-null     object
 4   Email         50 non-null     object
 5   Gov           60 non-null     object
 6   City          60 non-null     object
 7   Address       60 non-null     object
dtypes: object(8)
memory usage: 3.9+ KB


In [ ]:
CustomersDf.isna().sum()[CustomersDf.isna().sum() > 1]

CustomerID    20
Gender         6
Phone         10
Email         10
dtype: int64

#**Customer Name**


*   We have a Mix of Arabic & English names


      1.  Either we hard code it
      2.   List item





In [ ]:
CustomersDf["CustomerName"].unique()

array(['Alaa Fathy', 'Mohamed Ali', 'Youssef Nasr', 'علاء إبراهيم',
       'يوسف علي', 'محمد نصر', 'محمد فتحي', 'Karim Fathy', 'Alaa Nasr',
       'علاء علي', 'Dina Farouk', 'Nour Fathy', 'علاء راشد',
       'Karim Rashid', 'نور إبراهيم', 'Alaa Ibrahim', 'Sara Nasr',
       'Dina Mahmoud', 'Hany Ibrahim', 'منى علي', 'هاني راشد',
       'يوسف فاروق', 'Nour Nasr', 'سارة فاروق', 'Mohamed Farouk',
       'دينا حسن', 'كريم علي', 'منى فتحي', 'Youssef Hassan',
       'Hany Hassan', 'هاني محمود', 'Dina Rashid', 'Dina Ali',
       'دينا إبراهيم', 'Sara Fathy', 'Dina Hassan', 'Mohamed Ibrahim',
       'علاء نصر', 'سارة نصر', 'Hany Farouk', 'علاء فتحي', 'نور علي',
       'Hany Rashid', 'نور محمود', 'Ahmed Ali', 'كريم فاروق', 'Sara Ali',
       'Karim Farouk', 'Alaa Farouk', 'سارة حسن'], dtype=object)

In [ ]:
translator = Translator()
translated = translator.translate('سارة إبراهيم', src='ar', dest='en')
print(translated.text)

Sarah Ibrahim


#**After seeing the result on the sample case I'll apply it to the customer name column**

In [ ]:
CustomersDf['CustomerName'] = CustomersDf['CustomerName'].apply(
    lambda x: translator.translate(x, src='ar', dest='en').text if isinstance(x, str) else x
)

In [ ]:
CustomersDf["CustomerName"].unique()

array(['Alaa Fathi', 'Mohamed Ali', 'Youssef Nasr', 'Alaa Ibrahim',
       'Yusuf Ali', 'Muhammad Nasr', 'Muhammad Fathi', 'Karim Fathi',
       'Alaa Nasr', 'Alaa Ali', 'Dina Farouk', 'Nour Fathi',
       'Alaa Rashid', 'Karim Rashid', 'Nour Ibrahim', 'Sarah Nasser',
       'Dina Mahmoud', 'Hany Ibrahim', 'Mona Ali', 'Hani Rashid',
       'Youssef Farouk', 'Nour Nasser', 'Sarah Farouk', 'Mohamed Farouk',
       'Dina Hassan', 'Karim Ali', 'Mona Fathi', 'Youssef Hassan',
       'Hany Hassan', 'Hani Mahmoud', 'Dina Rashid', 'Dina Ali',
       'Dina Ibrahim', 'Sarah Fathi', 'Mohamed Ibrahim', 'Sarah Nasr',
       'Hany Farouk', 'Nour Ali', 'Hany Rashid', 'Nour Mahmoud',
       'Ahmed Ali', 'Karim Farouk', 'Sara Ali', 'Alaa Farouk',
       'Sarah Hassan'], dtype=object)

#**Now we Don't have any Arabic Entries in the Customer Name**

#**Gender**

In [ ]:
CustomersDf["Gender"].unique()

array(['male', 'female', nan], dtype=object)

In [ ]:
(CustomersDf['Gender'].astype(str).str.strip().str.lower() == 'nan').sum()

np.int64(6)

In [ ]:
CustomersDf['Gender'] = CustomersDf['Gender'].astype(str).str.strip().str.lower().map({'m':'male', 'f':'female', 'ذكر':'male', 'أنثى':'female'})

In [ ]:
CustomersDf["Gender"].unique()


array([nan], dtype=object)

In [ ]:
genderNulls=CustomersDf[CustomersDf.isna().any(axis=1)]['CustomerName']
print(genderNulls)

0          Alaa Fathi
1         Mohamed Ali
2        Youssef Nasr
3        Alaa Ibrahim
4           Yusuf Ali
5       Muhammad Nasr
6      Muhammad Fathi
7         Karim Fathi
8           Alaa Nasr
9            Alaa Ali
10        Dina Farouk
11         Nour Fathi
12        Alaa Rashid
13       Karim Rashid
14       Nour Ibrahim
15       Alaa Ibrahim
16       Sarah Nasser
17       Dina Mahmoud
18       Hany Ibrahim
19           Mona Ali
20        Hani Rashid
21     Youssef Farouk
22        Nour Nasser
23       Sarah Farouk
24     Mohamed Farouk
25        Dina Hassan
26          Karim Ali
27         Mona Fathi
28     Youssef Hassan
29        Hany Hassan
30       Hani Mahmoud
31        Dina Rashid
32         Nour Fathi
33           Dina Ali
34       Dina Ibrahim
35        Sarah Fathi
36        Mohamed Ali
37        Dina Hassan
38    Mohamed Ibrahim
39          Alaa Nasr
40       Youssef Nasr
41        Dina Farouk
42         Sarah Nasr
43        Hany Farouk
44      Muhammad Nasr
45        

#**I'll fix the gender by myself as I can get it from the name**

In [ ]:
CustomersDf['Gender'] = CustomersDf['CustomerName'].apply(lambda x: 'female' if any(f in x.lower() for f in ['dina', 'sarah', 'sara', 'mona', 'nour']) else 'male')

#**All the Nulls in the Gender Column and the inconsistancies are fixed**

In [ ]:
CustomersDf["Gender"].isna().sum()

np.int64(0)

#**Invalid or partial Email**

In [ ]:
CustomersDf["Email"].unique()

array(['alaa.fathyexample.com', 'mohamed.ali@example.com',
       'youssef.nasr@gmail', 'علاء.إبراهyمexample.com',
       'yوسف.علy@gmail.com', nan, 'محمد.فتحy@example.com',
       'alaa.nasr@example.com', 'علاء.علyexample.com',
       'dina.farouk@gmail.com', 'nour.fathy@example.com',
       'karim.rashid@example.com', 'نور.إبراهyم@gmail.com',
       'alaa.ibrahim@gmail.com', 'sara.nasr@gmail',
       'dina.mahmoud@gmail.com', 'هانy.راشد@gmail.com',
       'yوسف.فاروق@gmail.com', 'nour.nasr@gmail', 'سارة.فاروق@gmail',
       'mohamed.faroukexample.com', 'دyنا.حسن@example.com',
       'كرyم.علy@gmail', 'منy.فتحyexample.com',
       'youssef.hassan@gmail.com', 'هانy.محمود@example.com',
       'dina.rashid@gmail', 'nour.fathy@gmail.com', 'dina.aliexample.com',
       'دyنا.إبراهyم@example.com', 'sara.fathy@gmail',
       'dina.hassan@gmail.com', 'mohamed.ibrahim@gmail', 'علاء.نصر@gmail',
       'youssef.nasr@example.com', 'سارة.نصر@example.com',
       'hany.farouk@example.com', 'محمد.نص

#**Function For Adjusting Email Formate**

**Applying to DF**

In [ ]:
CustomersDf['Email_Cleaned'] = CustomersDf['Email'].apply(smart_email_cleaner)

In [ ]:
CustomersDf["Email_Cleaned"].unique()

array(['Alaa. fathy@example.com', 'mohamed.ali@example.com',
       'youssef.nasr@gmail', 'Alaa. Ibrahim@example.com',
       'yusf. Ali@gmail.com', None, 'Muhammad.Fathy@example.com',
       'alaa.nasr@example.com', 'Alaa. Ali@example.com',
       'dina. farouk@gmail.com', 'nour.fathy@example.com',
       'karim. rashid@example.com', 'Nour. Ibrahim@gmail.com',
       'alaa. ibrahim@gmail.com', 'sara.nasr@gmail',
       'in. mahmoud@gmail.com', 'Hany.Rashed@gmail.com',
       'Youssef Farouk@gmail.com', 'nour.nasr@gmail',
       'Sarah Farouk@gmail', 'mohamed.farouk@example.com',
       'dyna.hassan@example.com', 'Thank you@gmail',
       'My. Fathy@example.com', 'youssef.hassan@gmail.com',
       'Hany. Mahmoud@example.com', 'dina.rashid@gmail',
       'nour.fathy@gmail.com', 'dina.ali@example.com',
       'Dr. Ibrahim@example.com', 'sara. fathy@gmail',
       'dina. hassan@gmail.com', 'mohamed. ibrahim@gmail',
       'Alaa Nasr@gmail', 'youssef.nasr@example.com',
       'Sarah Nasr@e

In [ ]:
#SalesDf['Email_Cleaned'] = SalesDf['Email_Cleaned'].apply(
#    lambda x: str(x).replace(' ', '').strip().lower() if isinstance(x, str) else x
#)

In [ ]:
CustomersDf['Email_Cleaned'] = CustomersDf['Email_Cleaned'].apply(lambda x: (str(x).replace(' ', '').lower() + '.com') if isinstance(x, str) and '@' in x and '.com' not in x else str(x).replace(' ', '').lower() if isinstance(x, str) else x)

In [ ]:
CustomersDf["Email"]=CustomersDf["Email_Cleaned"]

In [ ]:
CustomersDf["Email"]

0         alaa.fathy@example.com
1        mohamed.ali@example.com
2         youssef.nasr@gmail.com
3       alaa.ibrahim@example.com
4             yusf.ali@gmail.com
5                           None
6     muhammad.fathy@example.com
7                           None
8          alaa.nasr@example.com
9           alaa.ali@example.com
10         dina.farouk@gmail.com
11        nour.fathy@example.com
12                          None
13      karim.rashid@example.com
14        nour.ibrahim@gmail.com
15        alaa.ibrahim@gmail.com
16           sara.nasr@gmail.com
17          in.mahmoud@gmail.com
18                          None
19                          None
20         hany.rashed@gmail.com
21       yousseffarouk@gmail.com
22           nour.nasr@gmail.com
23         sarahfarouk@gmail.com
24    mohamed.farouk@example.com
25       dyna.hassan@example.com
26            thankyou@gmail.com
27          my.fathy@example.com
28      youssef.hassan@gmail.com
29                          None
30      hany.mahmoud@example.com
31         dina.rashid@gmail.com
32          nour.fathy@gmail.com
33          dina.ali@example.com
34        dr.ibrahim@example.com
35          sara.fathy@gmail.com
36       mohamed.ali@example.com
37         dina.hassan@gmail.com
38     mohamed.ibrahim@gmail.com
39            alaanasr@gmail.com
40      youssef.nasr@example.com
41                          None
42         sarahnasr@example.com
43       only.farouk@example.com
44        muhammadnasr@gmail.com
45          alaa.fathy@gmail.com
46          nour.ali@example.com
47            thankyou@gmail.com
48       only.rashid@example.com
49        alaarashid@example.com
50                          None
51        nour.mahmoud@gmail.com
52         ahmed.ali@example.com
53                          None
54            sara.ali@gmail.com
55      karim.farouk@example.com
56      karim.rashid@example.com
57       dina.hassan@example.com
58         alaa.farouk@gmail.com
59                          None
Name: Email, dtype: object

#**ALL the Emails Are Consistant and Valide**

# **Phone Normalization**
**irregular Phone formats (+20, spaced, missing)**

In [ ]:
CustomersDf["Phone"].unique()

array(['01232297072', '01288174504', '+20 01142158168', '+2001247509496',
       '012 5369 1252', nan, '+2001038405950', '+2001226521948',
       '+20 01125527644', '+20 01166867849', '010 7708 6404',
       '011 4912 3810', '01231366552', '012 8355 2661', '+2001129437798',
       '010 5772 5656', '01260381528', '+2001015253706', '01198336655',
       '01223288780', '+2001161241954', '+20 01130601268',
       '+2001129640030', '+2001226960874', '+20 01177408068',
       '01242217185', '012 1097 8481', '+20 01078945564',
       '+20 01010691813', '010 7317 6107', '+2001058215939',
       '011 2594 4033', '+2001211831931', '+2001026168027',
       '010 5512 1027', '01015488046', '01099784619', '+2001081612882',
       '01277628209', '+20 01280474132', '01046297151', '+20 01249580182',
       '+20 01139613670', '+2001033637158', '011 4291 7412',
       '+20 01082950883', '+20 01057420453', '01058384619',
       '+20 01177008305', '+2001247376435', '+20 01012302458'],
      dtype=object)

In [ ]:
CustomersDf['Phone'] = CustomersDf['Phone'].apply(lambda x: str(x).replace('+20', '').replace(' ','').strip() if pd.notnull(x) else x)

In [ ]:
CustomersDf["Phone"].unique()

array(['01232297072', '01288174504', '01142158168', '01247509496',
       '01253691252', nan, '01038405950', '01226521948', '01125527644',
       '01166867849', '01077086404', '01149123810', '01231366552',
       '01283552661', '01129437798', '01057725656', '01260381528',
       '01015253706', '01198336655', '01223288780', '01161241954',
       '01130601268', '01129640030', '01226960874', '01177408068',
       '01242217185', '01210978481', '01078945564', '01010691813',
       '01073176107', '01058215939', '01125944033', '01211831931',
       '01026168027', '01055121027', '01015488046', '01099784619',
       '01081612882', '01277628209', '01280474132', '01046297151',
       '01249580182', '01139613670', '01033637158', '01142917412',
       '01082950883', '01057420453', '01058384619', '01177008305',
       '01247376435', '01012302458'], dtype=object)

#**All phone numbers are now without prefix**

#**Inconsistent geography (Arabic/English/typos)**

In [ ]:
CustomersDf["Gov"].unique()

array(['Sharqia', 'gharbia', 'LUXOR', 'luxor', 'Dakahlia', 'SOUTH SINAI',
       'الجيزة', 'Alexandria', 'Aswan', 'الإسكندرية', 'qalyubia',
       'Qalyubia', 'Asyut', 'Giza', 'sharqia', 'dakahlia', 'Al Red Sea',
       'CAIRO', 'أسيوط', 'giza', 'alexandria', 'الغربية', 'cairo',
       'Al Qalyubia', 'DAKAHLIA', 'القليوبية', 'aswan', 'الدقهلية',
       'ASWAN', 'أسوان', 'Gharbia', 'QALYUBIA', 'الأقصر', 'ASYUT',
       'SHARQIA', 'GHARBIA', 'South Sinai', 'Al Alexandria'], dtype=object)

In [ ]:
CustomersDf['Gov'] = CustomersDf['Gov'].apply(
    lambda x: translator.translate(x, src='ar', dest='en').text if isinstance(x, str) else x)

In [ ]:
CustomersDf["Gov"].unique()

array(['Sharqia', 'gharbia', 'LUXOR', 'luxor', 'Dakahlia', 'SOUTH SINAI',
       'Giza', 'Alexandria', 'Aswan', 'qalyubia', 'Qalyubia', 'Asyut',
       'sharqia', 'Al Red Sea', 'CAIRO', 'Assiut', 'giza', 'alexandria',
       'Western', 'cairo', 'Al Qalyubia', 'aswan', 'ASWAN', 'Gharbia',
       'QALYUBIA', 'Luxor', 'ASYUT', 'SHARQIA', 'GHARBIA', 'South Sinai',
       'Al Alexandria'], dtype=object)

In [ ]:
CustomersDf = CustomersDf.applymap(
    lambda x: str(x).lower().replace('eastern', 'gharbia').replace('western','sharqia').replace('gizah','giza').replace('asyut','assiut').removeprefix('al ') if not str(x).lower().startswith('alexandria') else str(x).lower()
     if isinstance(x, str) else x
)


/tmp/ipython-input-1056832121.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  CustomersDf = CustomersDf.applymap(


#**The Inconsistancies in the Governorate Column are Fixed**

In [ ]:
CustomersDf["Gov"].unique()

array(['sharqia', 'gharbia', 'luxor', 'dakahlia', 'south sinai', 'giza',
       'alexandria', 'aswan', 'qalyubia', 'assiut', 'red sea', 'cairo'],
      dtype=object)

#**multi-line addresses (embedded newlines) & Dealing with arabic words**

In [ ]:
CustomersDf["Address"].unique()

array(['8 street tahrir', '134 شارع tahrir', '26 street tahrir',
       '106 street النصر', '90 شارع tahrir', '78 شارع النصر',
       '84 شارع tahrir', '148 street tahrir', '30 شارع النصر',
       '57 street tahrir', '82 شارع النصر', '58 street النصر',
       '194 شارع tahrir', '182 شارع tahrir', '186 street النصر',
       '39 شارع النصر', '3 street النصر', '147 street النصر',
       '169 street النصر', '193 street النصر', '137 شارع النصر',
       '155 street tahrir', '53 street tahrir', '137 street tahrir',
       '32 street tahrir', '89 شارع tahrir', '200 street tahrir',
       '163 street tahrir', '109 street tahrir', '156 street tahrir',
       '87 street النصر', '119 شارع النصر', '51 street النصر',
       '69 street tahrir', '191 شارع النصر', '193 شارع tahrir',
       '3 شارع tahrir', '64 شارع النصر', '5 street النصر',
       '29 شارع tahrir', '72 street tahrir', '200 شارع النصر',
       '181 شارع النصر', '12 street النصر', '48 street النصر',
       '44 street النصر', '155 شارع ta

In [ ]:
CustomersDf['Address'] = CustomersDf['Address'].apply(lambda x: translator.translate(str(x).replace('\n', ' '), src='ar', dest='en').text if any('\u0600' <= c <= '\u06FF' for c in str(x)) else str(x).replace('\n', ' '))

#**Final Address Result**

In [ ]:
CustomersDf['Address'].unique()

array(['8 street tahrir', '134 Tahrir Street', '26 street tahrir',
       '106 street victory', '90 Tahrir Street', '78 Al-Nasr Street',
       '84 Tahrir Street', '148 street tahrir', '30 Al-Nasr Street',
       '57 street tahrir', '82 Al-Nasr Street', '58 street victory',
       '194 Tahrir Street', '182 Tahrir Street', '186 street Al-Nasr',
       '39 Al-Nasr Street', '3 street victory', '147 street Al-Nasr',
       '169 street victory', '193 street victory', '137 Al-Nasr Street',
       '155 street tahrir', '53 street tahrir', '137 street tahrir',
       '32 street tahrir', '89 Tahrir Street', '200 street tahrir',
       '163 street tahrir', '109 street tahrir', '156 street tahrir',
       '87 street victory', '119 Al-Nasr Street', '51 street victory',
       '69 street tahrir', '191 Al-Nasr Street', '193 Tahrir Street',
       '3 Tahrir Street', '64 Al-Nasr Street', '5 street victory',
       '29 Tahrir Street', '72 street tahrir', '200 Al-Nasr Street',
       '181 Al-Nasr Street'

# **Customer ID**

In [ ]:
CustomersDf["CustomerID"].unique()

array(['c1000', 'c1001', 'cus-502', 'cus-503', 'nan', 'c1006', 'c1008',
       'cus-511', 'cus-512', 'c1013', 'c1015', 'cus-517', 'cus-518',
       'c1019', 'cus-520', 'c1021', 'cus-522', 'cus-523', 'c1024',
       'c1025', 'cus-527', 'cus-530', 'c1031', 'cus-532', 'cus-533',
       'c1034', 'c1037', 'c1038', 'c1039', 'c1041', 'c1042', 'c1043',
       'c1046', 'c1047', 'c1049', 'cus-550', 'c1051', 'c1052', 'c1053',
       'cus-557', 'cus-559'], dtype=object)

In [ ]:
CustomersDf['CustomerID'] = CustomersDf['CustomerID'].apply(lambda x: x.replace('cus-', 'c') if isinstance(x, str) and x.startswith('cus-') else x)


In [ ]:
CustomersDf["CustomerID"].unique()

array(['c1000', 'c1001', 'c502', 'c503', 'nan', 'c1006', 'c1008', 'c511',
       'c512', 'c1013', 'c1015', 'c517', 'c518', 'c1019', 'c520', 'c1021',
       'c522', 'c523', 'c1024', 'c1025', 'c527', 'c530', 'c1031', 'c532',
       'c533', 'c1034', 'c1037', 'c1038', 'c1039', 'c1041', 'c1042',
       'c1043', 'c1046', 'c1047', 'c1049', 'c550', 'c1051', 'c1052',
       'c1053', 'c557', 'c559'], dtype=object)

#**Final Dataset Results**

In [ ]:
CustomersDf.head()

CustomerID  CustomerName Gender        Phone                     Email  \
0      c1000    alaa fathi   male  01232297072    alaa.fathy@example.com   
1      c1001   mohamed ali   male  01288174504   mohamed.ali@example.com   
2    cus-502  youssef nasr   male  01142158168    youssef.nasr@gmail.com   
3    cus-503  alaa ibrahim   male  01247509496  alaa.ibrahim@example.com   
4        nan     yusuf ali   male  01253691252        yusf.ali@gmail.com   

        Gov             City             Address             Email_Cleaned  
0   sharqia            tanta     8 street tahrir    alaa.fathy@example.com  
1   gharbia  sharm el sheikh   134 Tahrir Street   mohamed.ali@example.com  
2     luxor            tanta    26 street tahrir    youssef.nasr@gmail.com  
3     luxor  sharm el sheikh  106 street victory  alaa.ibrahim@example.com  
4  dakahlia        nasr city    90 Tahrir Street        yusf.ali@gmail.com

#**Governorates_Lookup_Noise**

# **Accessing Sales Orders Raw Sheet**

---



In [ ]:
GovDf = dataSheets.parse('Governorates_Lookup_Noise')

# **Primary Inspection DF head,tail**




In [ ]:
GovDf.head()

GovName  AltName   Latitude  Longitude
0  Al Cairo  القاهرة  30.006956  31.203068
1     Cairo    Cairo    30,0444    31,2357
2   القاهرة    Cairo  30.025546  31.257867
3    الجيزة     Giza    31.2089    30.0131
4      giza      NaN    30,0131    31,2089

In [ ]:
GovDf.tail()

GovName       AltName   Latitude  Longitude
31       red sea  البحر الأحمر  27.249584  33.790434
32  البحر الأحمر           NaN    33.8116    27.2579
33   south sinai    جنوب سيناء     33.975    28.5426
34    جنوب سيناء    جنوب سيناء  28.503624  33.981709
35    جنوب سيناء    جنوب سيناء    28,5426     33,975


**Gov Columns**


In [ ]:
GovDf.columns

Index(['GovName', 'AltName', 'Latitude', 'Longitude'], dtype='object')

**Checking Nulls and Data Types**

In [ ]:
GovDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   GovName    36 non-null     object
 1   AltName    26 non-null     object
 2   Latitude   35 non-null     object
 3   Longitude  35 non-null     object
dtypes: object(4)
memory usage: 1.3+ KB


In [ ]:
GovDf['GovName'].unique()

array(['Al Cairo', 'Cairo', 'القاهرة', 'الجيزة', 'giza', 'alexandria',
       'الإسكندرية', 'Al Asyut', 'أسيوط', 'أسوان', 'ASWAN', 'qalyubia',
       'القليوبية', 'QALYUBIA', 'Sharqia', 'sharqia', 'الدقهلية',
       'Al Dakahlia', 'dakahlia', 'الغربية', 'gharbia', 'Luxor', 'الأقصر',
       'LUXOR', 'Red Sea', 'red sea', 'البحر الأحمر', 'south sinai',
       'جنوب سيناء'], dtype=object)

In [ ]:
governorateMapFinal = {
    'Al Cairo': 'Cairo', 'Cairo': 'Cairo', 'القاهرة': 'Cairo',
    'الجيزة': 'Giza', 'giza': 'Giza',
    'alexandria': 'Alexandria', 'الإسكندرية': 'Alexandria',
    'Al Asyut': 'Asyut', 'أسيوط': 'Asyut',
    'أسوان': 'Aswan', 'ASWAN': 'Aswan',
    'qalyubia': 'Qalyubia', 'القليوبية': 'Qalyubia', 'QALYUBIA': 'Qalyubia',
    'Sharqia': 'Sharqia', 'sharqia': 'Sharqia',
    'الدقهلية': 'Dakahlia', 'Al Dakahlia': 'Dakahlia', 'dakahlia': 'Dakahlia',
    'الغربية': 'Gharbia', 'gharbia': 'Gharbia',
    'Luxor': 'Luxor', 'الأقصر': 'Luxor', 'LUXOR': 'Luxor',
    'Red Sea': 'Red Sea', 'red sea': 'Red Sea', 'البحر الأحمر': 'Red Sea',
    'south sinai': 'South Sinai', 'جنوب سيناء': 'South Sinai'
}

In [ ]:
GovDf['GovName'] = GovDf['GovName'].map(governorateMapFinal)

In [ ]:
GovDf['GovName'].unique()

array(['Cairo', 'Giza', 'Alexandria', 'Asyut', 'Aswan', 'Qalyubia',
       'Sharqia', 'Dakahlia', 'Gharbia', 'Luxor', 'Red Sea',
       'South Sinai'], dtype=object)

#**ALT Name**

In [ ]:
GovDf['AltName'].unique()

array(['القاهرة', 'Cairo', 'Giza', nan, 'الجيزة', 'Alexandria', 'Asyut',
       'أسوان', 'Aswan', 'القليوبية', 'الشرقية', 'Dakahlia', 'Gharbia',
       'الأقصر', 'Luxor', 'البحر الأحمر', 'جنوب سيناء'], dtype=object)

In [ ]:
governorateMapFinalAR= {
    'Cairo': 'القاهرة',
    'Giza': 'الجيزة',
    'Alexandria': 'الإسكندرية',
    'Asyut': 'أسيوط',
    'Aswan': 'أسوان',
    'Qalyubia': 'القليوبية',
    'Sharqia': 'الشرقية',
    'Dakahlia': 'الدقهلية',
    'Gharbia': 'الغربية',
    'Luxor': 'الأقصر',
    'Red Sea': 'البحر الأحمر',
    'South Sinai': 'جنوب سيناء'
}

In [ ]:
GovDf['AltName'] = GovDf['GovName'].map(governorateMapFinalAR)

In [ ]:
GovDf['AltName']

0          القاهرة
1          القاهرة
2          القاهرة
3           الجيزة
4           الجيزة
5           الجيزة
6       الإسكندرية
7       الإسكندرية
8       الإسكندرية
9            أسيوط
10           أسيوط
11           أسيوط
12           أسوان
13           أسوان
14           أسوان
15       القليوبية
16       القليوبية
17       القليوبية
18         الشرقية
19         الشرقية
20         الشرقية
21        الدقهلية
22        الدقهلية
23        الدقهلية
24         الغربية
25         الغربية
26         الغربية
27          الأقصر
28          الأقصر
29          الأقصر
30    البحر الأحمر
31    البحر الأحمر
32    البحر الأحمر
33      جنوب سيناء
34      جنوب سيناء
35      جنوب سيناء
Name: AltName, dtype: object

#**Converted lat/long values from string to float, fixed decimal commas, flagged coordinates, and swapped the wrong latitude and longitude**

In [ ]:
GovDf[['Latitude', 'Longitude']] = GovDf[['Latitude', 'Longitude']].applymap(lambda x: float(str(x).replace(',', '.')) if pd.notna(x) else pd.NA)

/tmp/ipython-input-836132967.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  GovDf[['Latitude', 'Longitude']] = GovDf[['Latitude', 'Longitude']].applymap(lambda x: float(str(x).replace(',', '.')) if pd.notna(x) else pd.NA)


In [ ]:
GovDf[['Latitude', 'Longitude']]

Latitude   Longitude
0   30.006956   31.203068
1     30.0444     31.2357
2   30.025546   31.257867
3     31.2089     30.0131
4     30.0131     31.2089
5     30.0131     31.2089
6   31.186058   29.906472
7   31.211039   29.912043
8        <NA>        <NA>
9   27.205886   31.225853
10    31.1837       27.18
11  27.160585   31.162023
12  24.072223   32.921286
13  24.039318   32.935966
14  24.080482   32.889265
15  30.434105   31.288831
16  30.382448   31.323213
17   30.46571   31.307127
18  30.745498   31.700818
19  30.724404   31.688679
20  30.718141   31.734656
21  31.005688   31.389044
22    31.0409     31.3785
23  30.992134   31.405597
24  30.916015   31.140935
25    31.1829     30.8686
26  30.868665    31.18622
27   88.03547  108.931614
28  25.714869   32.624039
29  25.730202   32.629397
30  27.289539   33.761758
31  27.249584   33.790434
32    33.8116     27.2579
33     33.975     28.5426
34  28.503624   33.981709
35    28.5426      33.975

In [ ]:
GovDf['LatLong_Swapped'] = GovDf.apply(lambda row: pd.notna(row['Latitude']) and pd.notna(row['Longitude']) and -90 <= row['Latitude'] <= 90 and -180 <= row['Longitude'] <= 180, axis=1)

In [ ]:
GovDf['LatLong_Swapped'].value_counts()

LatLong_Swapped
True     35
False     1
Name: count, dtype: int64

In [ ]:
GovDf.loc[GovDf['LatLong_Swapped'], ['Latitude', 'Longitude']] = GovDf.loc[GovDf['LatLong_Swapped'], ['Longitude', 'Latitude']].values

#Saving into Excel again

In [ ]:
!pip install xlsxwriter

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 175.3/175.3 kB 5.5 MB/s eta 0:00:00


In [ ]:
SalesDf.columns

Index(['OrderID', 'OrderDate', 'DeliveryDate', 'CustomerID', 'CustomerName',
       'Gender', 'Phone', 'Email', 'Governorate', 'City', 'Address',
       'Latitude', 'Longitude', 'ProductSKU', 'ProductName', 'Category',
       'UnitPrice', 'Quantity', 'Subtotal', 'Discount', 'TotalAmount',
       'Currency', 'PaymentMethod', 'PaymentStatus', 'ShipperName',
       'ShippingCost', 'SalesRep', 'Channel', 'ReturnFlag', 'ReturnDate',
       'Status', 'Notes', 'LatLong_Swapped', 'Email_Cleaned'],
      dtype='object')

In [ ]:
SalesDf = SalesDf.drop(columns=['Email_Cleaned'])


In [ ]:
SalesDf = SalesDf.drop(columns=['LatLong_Swapped'])


In [ ]:
CustomersDf.columns

Index(['CustomerID', 'CustomerName', 'Gender', 'Phone', 'Email', 'Gov', 'City',
       'Address'],
      dtype='object')

In [ ]:
CustomersDf = CustomersDf.drop(columns=['Email_Cleaned'])


In [ ]:
GovDf.columns

Index(['GovName', 'AltName', 'Latitude', 'Longitude', 'Swap', 'Swapped',
       'LatLong_Swapped'],
      dtype='object')

In [ ]:
GovDf = GovDf.drop(columns=['Swap', 'Swapped','LatLong_Swapped'])


In [ ]:
with pd.ExcelWriter("cleaned_data_.xlsx", engine='xlsxwriter') as writer:
    SalesDf.to_excel(writer, sheet_name='Sales_Orders', index=False)
    CustomersDf.to_excel(writer, sheet_name='Customers', index=False)
    ProductDf.to_excel(writer, sheet_name='Products', index=False)
    GovDf.to_excel(writer, sheet_name='Gov_LookUp', index=False)


In [ ]:
SalesDf.to_csv("Sales_Orders.csv", index=False)
CustomersDf.to_csv("Customers.csv", index=False)
ProductDf.to_csv("Products.csv", index=False)
GovDf.to_csv("Gov_LookUp.csv", index=False)

In [ ]:
ProductDf.columns

Index(['SKU', 'ProductName', 'Category'], dtype='object')

#**Fixing Some Data Types**

In [ ]:
SalesDf.head()


OrderID OrderDate         DeliveryDate CustomerID  CustomerName  Gender  \
36   so85088       nat  2024-01-07 00:00:00        nan  karim hassan    male   
37   so17567       nat  2024-06-04 00:00:00        nan   alaa farouk    male   
110  so27394       nat  2024-04-03 00:00:00        nan   ahmed fathi  female   
25   so77839       nat  2024-05-18 00:00:00       c576  mona ibrahim    male   
138  so84130       nat  2024-05-01 00:00:00       c576   alaa farouk    male   

           Phone                      Email Governorate     City  ...  \
36   01192493631  generous.hassan@gmail.com  alexandria  montaza  ...   
37   01073360324    alaa.farouk@example.com     sharqia  zagazig  ...   
110  01264406548    ahmed.fathy@example.com     red sea   safaga  ...   
25   01158895439     mona.ibrahim@gmail.com       luxor     esna  ...   
138  01278495121     alaafarouk@example.com  alexandria   smouha  ...   

    PaymentMethod  PaymentStatus  ShipperName ShippingCost      SalesRep  \
36          Fawry           paid       aramex         25.0  karim rashid   
37     mastercard         unpaid        fedex         75.0     ahmed ali   
110          visa         unpaid       aramex          0.0  karim rashid   
25           visa        pending   Egypt Post          0.0  karim rashid   
138         meeza         unpaid       aramex         25.0  karim rashid   

       Channel  ReturnFlag  ReturnDate      Status  \
36   tel-sales          no         NaT    returned   
37       store         yes  2024-06-22         new   
110      e-com          no         NaT         new   
25       e-com          no         NaT  processing   
138  tel-sales         yes  2024-05-01    returned   

                             Notes  
36   Old customer - extra discount  
37                              in  
110                             in  
25                              in  
138   customer asked for gift wrap  

[5 rows x 32 columns]

In [ ]:
dataSheets = pd.ExcelFile('/content/cleaned_data_.xlsx')
SalesDf = dataSheets.parse('Sales_Orders')

In [ ]:
CustomersDf= dataSheets.parse('Customers')
ProductDf= dataSheets.parse('Products')
GovDf= dataSheets.parse('Gov_LookUp')

#**Transforming the dates from objects to date formate**

In [ ]:
SalesDf['OrderDate'] = pd.to_datetime(SalesDf['OrderDate'], errors='coerce')


In [ ]:
SalesDf['DeliveryDate'] = pd.to_datetime(SalesDf['DeliveryDate'], errors='coerce')


In [ ]:
SalesDf['ReturnDate'] = pd.to_datetime(SalesDf['ReturnDate'], errors='coerce')

In [ ]:
SalesDf['DeliveryDate'].head()

0   2024-01-07
1   2024-06-04
2   2024-04-03
3   2024-05-18
4   2024-05-01
Name: DeliveryDate, dtype: datetime64[ns]

In [ ]:
SalesDf['OrderDate'] = pd.to_datetime(SalesDf['OrderDate'], errors='coerce')
SalesDf['OrderDate'] = SalesDf['OrderDate'].dt.strftime('%Y-%m-%d')


In [ ]:
SalesDf['DeliveryDate'] = pd.to_datetime(SalesDf['DeliveryDate'], errors='coerce')
SalesDf['DeliveryDate'] = SalesDf['DeliveryDate'].dt.strftime('%Y-%m-%d')

In [ ]:
SalesDf['ReturnDate'] = pd.to_datetime(SalesDf['ReturnDate'], errors='coerce')
SalesDf['ReturnDate'] = SalesDf['ReturnDate'].dt.strftime('%Y-%m-%d')

In [ ]:
SalesDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136 entries, 0 to 135
Data columns (total 32 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   OrderID        136 non-null    object 
 1   OrderDate      122 non-null    object 
 2   DeliveryDate   123 non-null    object 
 3   CustomerID     87 non-null     object 
 4   CustomerName   136 non-null    object 
 5   Gender         136 non-null    object 
 6   Phone          114 non-null    object 
 7   Email          105 non-null    object 
 8   Governorate    136 non-null    object 
 9   City           136 non-null    object 
 10  Address        136 non-null    object 
 11  Latitude       117 non-null    float64
 12  Longitude      117 non-null    float64
 13  ProductSKU     136 non-null    object 
 14  ProductName    136 non-null    object 
 15  Category       136 non-null    object 
 16  UnitPrice      136 non-null    float64
 17  Quantity       136 non-null    int64  
 18  Subtotal  

In [ ]:
CustomersDf['Phone'] = CustomersDf['Phone'].astype(str)

In [ ]:
CustomersDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   CustomerID    40 non-null     object
 1   CustomerName  60 non-null     object
 2   Gender        60 non-null     object
 3   Phone         60 non-null     object
 4   Email         60 non-null     object
 5   Gov           60 non-null     object
 6   City          60 non-null     object
 7   Address       60 non-null     object
dtypes: object(8)
memory usage: 3.9+ KB


In [ ]:
dataSheets = pd.ExcelFile('/content/EG_Retail_Data_Cleaned_.xlsx')
ProductDf = dataSheets.parse('Products')

In [ ]:
ProductDf

SKU              ProductName     Category
0   elec001  Laptop i7 16GB - لابتوب  electronics
1   elec001  LAPTOP I7 16GB - لابتوب  electronics
2   elec001  Labtop i7 16GB - لابتوب  electronics
3   elec002    Smartphone A55 - هاتف  electronics
4   elec002    Smartphone A55 - هاتف  electronics
5   elec002    Smartphone A55 - هاتف  electronics
6   home101  Electric Kettle - غلاية         home
7   home101  ELECTRIC KETTLE - غلاية         home
8   home101  Electric Kettle - غلاية         home
9   groc501     Organic زيت زيتون 1L      grocery
10  groc501     Organic زيت زيتون 1L      grocery
11  groc501     Organic زيت زيتون 1L      grocery
12  fash210     Men T-Shirt - تيشيرت      fashion
13  fash210     Men T-Shirt - تيشيرت      fashion
14  fash210     Men T-Shirt - تيشيرت      fashion
15  sprt301       FOOTBALL - كرة قدم       sports
16  sprt301       FOOTBALL - كرة قدم       sports
17  sprt301       Football - كرة قدم       sports
18  book900     Data Warehousing 101        books
19  book900     DATA WAREHOUSING 101        books
20  book900     DATA WAREHOUSING 101        books
21  elec003          BLUETOOTH سماعة  electronics
22  elec003          Bluetooth سماعة  electronics
23  elec003          Bluetooth سماعة  electronics
24  home102     Air Fryer 4L - قلاية         home
25  home102     Air Fryer 4L - قلاية         home
26  home102     AIR FRYER 4L - قلاية         home
27   toy777   PUZZLE 1000PCS - أحجية         toys
28   toy777   PUZZLE 1000PCS - أحجية         toys
29   toy777   Puzzle 1000pcs - أحجية         toys

In [ ]:
ProductDf = ProductDf.drop_duplicates(subset=['SKU'], keep='first').reset_index(drop=True)

In [ ]:
ProductDf

SKU              ProductName     Category
0  elec001  Laptop i7 16GB - لابتوب  electronics
1  elec002    Smartphone A55 - هاتف  electronics
2  home101  Electric Kettle - غلاية         home
3  groc501     Organic زيت زيتون 1L      grocery
4  fash210     Men T-Shirt - تيشيرت      fashion
5  sprt301       FOOTBALL - كرة قدم       sports
6  book900     Data Warehousing 101        books
7  elec003          BLUETOOTH سماعة  electronics
8  home102     Air Fryer 4L - قلاية         home
9   toy777   PUZZLE 1000PCS - أحجية         toys

In [ ]:
CustomersDf = CustomersDf.drop_duplicates(subset=['CustomerID'], keep='first').reset_index(drop=True)

In [ ]:
GovDf

GovName       AltName    Latitude  Longitude
0         Cairo       القاهرة   31.203068  30.006956
1         Cairo       القاهرة   31.235700  30.044400
2         Cairo       القاهرة   31.257867  30.025546
3          Giza        الجيزة   30.013100  31.208900
4          Giza        الجيزة   31.208900  30.013100
5          Giza        الجيزة   31.208900  30.013100
6    Alexandria    الإسكندرية   29.906472  31.186058
7    Alexandria    الإسكندرية   29.912043  31.211039
8    Alexandria    الإسكندرية         NaN        NaN
9         Asyut         أسيوط   31.225853  27.205886
10        Asyut         أسيوط   27.180000  31.183700
11        Asyut         أسيوط   31.162023  27.160585
12        Aswan         أسوان   32.921286  24.072223
13        Aswan         أسوان   32.935966  24.039318
14        Aswan         أسوان   32.889265  24.080482
15     Qalyubia     القليوبية   31.288831  30.434105
16     Qalyubia     القليوبية   31.323213  30.382448
17     Qalyubia     القليوبية   31.307127  30.465710
18      Sharqia       الشرقية   31.700818  30.745498
19      Sharqia       الشرقية   31.688679  30.724404
20      Sharqia       الشرقية   31.734656  30.718141
21     Dakahlia      الدقهلية   31.389044  31.005688
22     Dakahlia      الدقهلية   31.378500  31.040900
23     Dakahlia      الدقهلية   31.405597  30.992134
24      Gharbia       الغربية   31.140935  30.916015
25      Gharbia       الغربية   30.868600  31.182900
26      Gharbia       الغربية   31.186220  30.868665
27        Luxor        الأقصر  108.931614  88.035470
28        Luxor        الأقصر   32.624039  25.714869
29        Luxor        الأقصر   32.629397  25.730202
30      Red Sea  البحر الأحمر   33.761758  27.289539
31      Red Sea  البحر الأحمر   33.790434  27.249584
32      Red Sea  البحر الأحمر   27.257900  33.811600
33  South Sinai    جنوب سيناء   28.542600  33.975000
34  South Sinai    جنوب سيناء   33.981709  28.503624
35  South Sinai    جنوب سيناء   33.975000  28.542600

In [ ]:
with pd.ExcelWriter("EG_Retail_Data_Cleaned_Final_.xlsx", engine='xlsxwriter') as writer:
    SalesDf.to_excel(writer, sheet_name='Sales_Orders', index=False)
    CustomersDf.to_excel(writer, sheet_name='Customers', index=False)
    ProductDf.to_excel(writer, sheet_name='Products', index=False)
    GovDf.to_excel(writer, sheet_name='Gov_LookUp', index=False)

In [ ]:
print(3)

3


In [ ]:
SalesDf.isna().sum()

OrderID            0
OrderDate         14
DeliveryDate      13
CustomerID        49
CustomerName       0
Gender             0
Phone             22
Email             31
Governorate        0
City               0
Address            0
Latitude          19
Longitude         19
ProductSKU         0
ProductName        0
Category           0
UnitPrice          0
Quantity           0
Subtotal           0
Discount           0
TotalAmount        0
Currency           0
PaymentMethod      0
PaymentStatus      0
ShipperName        0
ShippingCost      14
SalesRep           0
Channel            0
ReturnFlag         0
ReturnDate       133
Status            14
Notes              0
dtype: int64

In [ ]:
sales=SalesDf
sales = sales.merge(CustomersDf[["CustomerName", "CustomerID"]], on="CustomerName", how="left", suffixes=("", "_from_customers"))
sales["CustomerID"] = sales["CustomerID"].fillna(sales["CustomerID_from_customers"])
sales = sales.drop(columns=["CustomerID_from_customers"])

In [ ]:
CustomersDf["CustomerName"] = CustomersDf["CustomerName"].str.lower().str.strip()

In [ ]:
null_customerid_rows = sales[sales["CustomerID"].isna()]

# Display them
print(null_customerid_rows)

     OrderID   OrderDate DeliveryDate CustomerID      CustomerName  Gender  \
0    so85088         NaN   2024-01-07        NaN      karim hassan    male   
1    so17567         NaN   2024-06-04        NaN       alaa farouk    male   
2    so27394         NaN   2024-04-03        NaN       ahmed fathi  female   
17   so46467  2024-06-26   2024-06-24        NaN      mohamed nasr  female   
25   so46408  2024-06-17   2024-06-18        NaN      mohamed nasr    male   
27   so48854  2024-06-15   2024-06-17        NaN       alaa hassan  female   
31   so15229  2024-06-13   2024-06-13        NaN     sarah mahmoud  female   
34   so17641  2024-06-06   2024-06-08        NaN      alaa mahmoud    male   
35   so63523  2024-06-02   2024-06-06        NaN         mona nasr  female   
37   so72204  2024-05-30   2024-06-13        NaN     sarah mahmoud  female   
38   so76619  2024-05-29          NaN        NaN     youssef fathi  female   
41   so80673  2024-05-28   2024-06-03        NaN    youssef rash

In [ ]:
for df in [sales, CustomersDf]:
    df["CustomerName"] = df["CustomerName"].str.lower().str.strip()
    df["Email"] = df["Email"].str.lower().str.strip()
sales = sales.merge(CustomersDf[["CustomerName", "CustomerID", "Email"]],
                    on="CustomerName", how="left", suffixes=("", "_from_name"))
sales = sales.merge(CustomersDf[["Email", "CustomerID"]],
                    on="Email", how="left", suffixes=("", "_from_email"))

sales["CustomerID"] = (
    sales["CustomerID"]
    .fillna(sales["CustomerID_from_name"])
    .fillna(sales["CustomerID_from_email"])
)

sales = sales.drop(columns=["CustomerID_from_name", "CustomerID_from_email"])


In [ ]:
sales["Email"].isna().sum()

np.int64(37)

In [ ]:
GovDf.isna().sum()

GovName      0
AltName      0
Latitude     1
Longitude    1
dtype: int64